---
title: "luad-v.2.0_druggability_2"
author: "Shaleigh Smith"
date: "5/21/2019"
output: html_document
---




---

Modified druggability Analysis 

---

```{r}

library(tidyverse)
library(dplyr)
library(rDGIdb)
library(plyr)
library(data.table)
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
library(tidyr)

```


---

```{r phos read in gene lists and merge}

##### 2 Group All Upregulated, filtered for DGIDB direct interaction list: Phosphoproteome #####

# Read in all phosphoproteome significant outliers 
# join all by gene name
# leave significance 
# make this into a heatmap


gender_all <- read.table("./comparison_results_1_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_up_female__male____sig_outliers.txt", header = TRUE, sep = "\t")
gender_all$GeneSymbol <- as.character(gender_all$GeneSymbol)
colnames(gender_all)[2] <- "gender_all"

region_all <- read.table("./comparison_results_1_dgidb_fdr_0.01/luad2_phos_dgidb_Region_up_asian__western____sig_outliers.txt", header = TRUE, sep = "\t")
region_all$GeneSymbol <- as.character(region_all$GeneSymbol)
colnames(region_all)[2] <- "region_all"

smoking_all <- read.table("./comparison_results_1_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_up_smoker__non-smoker____sig_outliers.txt", header = TRUE, sep = "\t")
smoking_all$GeneSymbol <- as.character(smoking_all$GeneSymbol)
colnames(smoking_all)[2] <- "smoking_all"

egfr_all <- read.table("./comparison_results_1_dgidb_fdr_0.01/luad2_phos_dgidb_EGFR_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
egfr_all$GeneSymbol <- as.character(egfr_all$GeneSymbol)
colnames(egfr_all)[2] <- "egfr_all"

kras_all <- read.table("./comparison_results_1_dgidb_fdr_0.01/luad2_phos_dgidb_KRAS_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
kras_all$GeneSymbol <- as.character(kras_all$GeneSymbol)
colnames(kras_all)[2] <- "kras_all"

tp53_all <- read.table("./comparison_results_1_dgidb_fdr_0.01/luad2_phos_dgidb_TP53_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
tp53_all$GeneSymbol <- as.character(tp53_all$GeneSymbol)
colnames(tp53_all)[2] <- "tp53_all"

egfr_female <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_EGFR_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_female$GeneSymbol <- as.character(egfr_female$GeneSymbol)
colnames(egfr_female)[2] <- "egfr_female"

egfr_male <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_EGFR_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_male$GeneSymbol <- as.character(egfr_male$GeneSymbol)
colnames(egfr_male)[2] <- "egfr_male"

kras_female <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_KRAS_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_female$GeneSymbol <- as.character(kras_female$GeneSymbol)
colnames(kras_female)[2] <- "kras_female"

kras_male <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_KRAS_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_male$GeneSymbol <- as.character(kras_male$GeneSymbol)
colnames(kras_male)[2] <- "kras_male"

tp53_female <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_TP53_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_female$GeneSymbol <- as.character(tp53_female$GeneSymbol)
colnames(tp53_female)[2] <- "tp53_female"

tp53_male <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Gender_TP53_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_male$GeneSymbol <- as.character(tp53_male$GeneSymbol)
colnames(tp53_male)[2] <- "tp53_male"

egfr_asian <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_EGFR_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_asian$GeneSymbol <- as.character(egfr_asian$GeneSymbol)
colnames(egfr_asian)[2] <- "egfr_asian"

egfr_western <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_EGFR_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_western$GeneSymbol <- as.character(egfr_western$GeneSymbol)
colnames(egfr_western)[2] <- "egfr_western"

gender_asian <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_Gender_up_asian_female_asian_male___sig_outliers.txt", header = TRUE, sep = "\t")
gender_asian$GeneSymbol <- as.character(gender_asian$GeneSymbol)
colnames(gender_asian)[2] <- "gender_asian"

region_female <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_Gender_up_asian_female_western_female___sig_outliers.txt", header = TRUE, sep = "\t")
region_female$GeneSymbol <- as.character(region_female$GeneSymbol)
colnames(region_female)[2] <- "region_female"

region_male <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_Gender_up_asian_male_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
region_male$GeneSymbol <- as.character(region_male$GeneSymbol)
colnames(region_male)[2] <- "region_male"

gender_western <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_Gender_up_western_female_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
gender_western$GeneSymbol <- as.character(gender_western$GeneSymbol)
colnames(gender_western)[2] <- "gender_western"

kras_asian <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_KRAS_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_asian$GeneSymbol <- as.character(kras_asian$GeneSymbol)
colnames(kras_asian)[2] <- "kras_asian"

kras_western <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_KRAS_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_western$GeneSymbol <- as.character(kras_western$GeneSymbol)
colnames(kras_western)[2] <- "kras_western"

tp53_asian <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_TP53_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_asian$GeneSymbol <- as.character(tp53_asian$GeneSymbol)
colnames(tp53_asian)[2] <- "tp53_asian"

tp53_western <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Region_TP53_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_western$GeneSymbol <- as.character(tp53_western$GeneSymbol)
colnames(tp53_western)[2] <- "tp53_western"

smoker_female <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_Gender_up_smoker_female_non-smoker_female___sig_outliers.txt", header = TRUE, sep = "\t")
smoker_female$GeneSymbol <- as.character(smoker_female$GeneSymbol)
colnames(smoker_female)[2] <- "smoker_female"

#smoker_male <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_Gender_up_smoker_male_non-smoker_male___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_male$GeneSymbol <- as.character(smoker_male$GeneSymbol)
#colnames(smoker_male)[2] <- "smoker_male"

#smoker_asian <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_Region_up_smoker_asian_non-smoker_asian___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_asian$GeneSymbol <- as.character(smoker_asian$GeneSymbol)
#colnames(smoker_asian)[2] <- "smoker_asian"

smoker_western <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_Region_up_smoker_western_non-smoker_western___sig_outliers.txt", header = TRUE, sep = "\t")
smoker_western$GeneSymbol <- as.character(smoker_western$GeneSymbol)
colnames(smoker_western)[2] <- "smoker_western"

smoker_egfr <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_EGFR_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
smoker_egfr$GeneSymbol <- as.character(smoker_egfr$GeneSymbol)
colnames(smoker_egfr)[2] <- "smoker_egfr"

#smoker_kras <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_KRAS_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_kras$GeneSymbol <- as.character(smoker_kras$GeneSymbol)
#colnames(smoker_kras)[2] <- "smoker_kras"

smoker_tp53 <- read.table("./comparison_results_2_dgidb_fdr_0.01/luad2_phos_dgidb_Smoking_TP53_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
smoker_tp53$GeneSymbol <- as.character(smoker_tp53$GeneSymbol)
colnames(smoker_tp53)[2] <- "smoker_tp53"

###

phos_dgidb <- full_join(egfr_all, egfr_female, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, egfr_male, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, egfr_asian, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, egfr_western, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, kras_all, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, kras_female, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, kras_male, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, kras_asian, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, kras_western, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, tp53_all, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, tp53_female, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, tp53_male, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, tp53_asian, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, tp53_western, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, region_all, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, region_female, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, region_male, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, gender_all, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, gender_asian, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, gender_western, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, smoking_all, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, smoker_female, by = "GeneSymbol")
#phos_dgidb <- full_join(phos_dgidb, smoker_male, by = "GeneSymbol")
#phos_dgidb <- full_join(phos_dgidb, smoker_asian, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, smoker_western, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, smoker_egfr, by = "GeneSymbol")
phos_dgidb <- full_join(phos_dgidb, smoker_tp53, by = "GeneSymbol")
#phos_dgidb <- full_join(phos_dgidb, smoker_kras, by = "GeneSymbol")

phos_dgidb[is.na(phos_dgidb)] <- 0

###

phos_dgidb_gene <- phos_dgidb[,-1]
rownames(phos_dgidb_gene) <- phos_dgidb[,1]
cohorts_all <- colnames(phos_dgidb_gene)
phos_dgidb_annotation <- HeatmapAnnotation(df = as.data.frame(cohorts_all),
                                       col = list(cohorts_all = c(egfr_all = "coral",
                                                              egfr_female = "dodgerblue4",
                                                              egfr_male = "dodgerblue1",
                                                              egfr_asian = "deepskyblue4",
                                                              egfr_western = "deepskyblue2",
                                                              kras_all = "coral",
                                                              kras_female = "mediumorchid4",
                                                              kras_male = "mediumorchid1",
                                                              kras_asian = "mediumpurple4",
                                                              kras_western = "mediumpurple1",
                                                              tp53_all = "coral",
                                                              tp53_female = "olivedrab4",
                                                              tp53_male = "olivedrab3",
                                                              tp53_asian = "darkgreen",
                                                              tp53_western = "springgreen4",
                                                              region_all = "coral",
                                                              region_female = "darkorange3",
                                                              region_male = "darkorange",
                                                              gender_all = "coral",
                                                              gender_asian = "darkgoldenrod3",
                                                              gender_western = "darkgoldenrod1",
                                                              smoking_all = "coral",
                                                              smoker_female = "lightsteelblue4",
                                                              #smoker_male = "lightsteelblue1",
                                                              #smoker_asian = "mistyrose4",
                                                              smoker_western = "mistyrose2",
                                                              smoker_egfr = "aquamarine4",
                                                              #smoker_kras = "aquamarine3",
                                                              smoker_tp53 = "aquamarine1")))

phos_dgidb_heat <- Heatmap(phos_dgidb_gene,
                         top_annotation = phos_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(3, "in"),
                                                     title = NULL),
                         col = colorRamp2(c(0, 1.0e-50, 0.0001, 0.01), 
                                          c("gray85", "red3", "orangered2", "oldlace")),
                         row_title = "Q Value",
                         row_title_gp = gpar(fontsize = 10))
  
phos_dgidb_heat_2 <- draw(phos_dgidb_heat, heatmap_legend_side = "left")

```


```{r phos dgidb query}

# DGIDB

# Filterd
druggable_dgidb <- rownames(phos_dgidb_gene)
length(druggable_dgidb) #63

# Query 
result_drug_dgidb <- queryDGIdb(druggable_dgidb)

```


```{r phos dgidb result summary}

## Result summary
# Score is the total number of source databases listing the interaction (including PubMed articles)
summary_drug_dgidb <- resultSummary(result_drug_dgidb) 
summary_drug_dgidb_score <- dplyr::select(summary_drug_dgidb, Gene, Score)

# Change to numeric and aggregate for total score
summary_drug_dgidb_score$Score <- as.numeric(summary_drug_dgidb_score$Score)
summary_drug_dgidb_score <- aggregate(. ~ Gene, data = summary_drug_dgidb_score, sum)

# Get log score
summary_drug_dgidb_score$Log_Score <- log(summary_drug_dgidb_score$Score)

# Order rows to match heatmap
target <- as.vector(rownames(phos_dgidb_gene))
summary_drug_dgidb_score <- summary_drug_dgidb_score[match(target, summary_drug_dgidb_score$Gene),]

# Label row names for heatmap
row.names(summary_drug_dgidb_score) <- target

# Replace NA with 0 for plotting
summary_drug_dgidb_score$Score[is.na(summary_drug_dgidb_score$Score) == TRUE] <- 0
summary_drug_dgidb_score$Log_Score[is.na(summary_drug_dgidb_score$Log_Score) == TRUE] <- 0


summary_drug_dgidb_score

```


```{r phos import interaction and aggregate}

# Import manual interactions then aggregate and separate into 1/0
interactions <- read.delim("./druggable/DGIDB_gene_drug_interactions_20190417_modified_dup_removed.txt", sep = "\t")
interactions <- interactions[interactions$gene_name %in% druggable_dgidb ,]

# Select gene name and interaction type only
interactions <- dplyr::select(interactions, gene_name, interaction_types)

# Aggregate to a list
interactions <- aggregate(interaction_types ~ gene_name, interactions, paste, collapse = ",")

# Order rows to match heatmap
target_int <- as.vector(druggable_dgidb)
interactions <- interactions[match(target_int, interactions$gene_name),]

interactions

```


```{r phos manual interactions data frame}

# Rename columns
colnames(interactions)[1:2] <- c("gene", "interaction_type")


# if interaction_type includes this exact match, put 1 else put 0 

interactions <- separate(interactions, interaction_type, 
                         into=c("int_1", "int_2", "int_3", "int_4", "int_5"), sep="," )
interactions$int_1[interactions$int_1==''] <- NA
interactions$int_1[is.na(interactions$int_1) == TRUE] <- 0
interactions$int_2[is.na(interactions$int_2) == TRUE] <- 0
interactions$int_3[is.na(interactions$int_3) == TRUE] <- 0
interactions$int_4[is.na(interactions$int_4) == TRUE] <- 0
interactions$int_5[is.na(interactions$int_5) == TRUE] <- 0

interactions

interactions$Antagonist <-ifelse((interactions$int_1 == "antagonist" |
                                interactions$int_2 == "antagonist" |
                                interactions$int_3 == "antagonist" |
                                interactions$int_4 == "antagonist" |
                                interactions$int_5 == "antagonist"), 1, 0)

interactions$Inhibitor <-ifelse((interactions$int_1 == "inhibitor" |
                                interactions$int_2 == "inhibitor" |
                                interactions$int_3 == "inhibitor" |
                                interactions$int_4 == "inhibitor" |
                                interactions$int_5 == "inhibitor"), 1, 0)

interactions$Agonist <-ifelse((interactions$int_1 == "agonist" |
                                interactions$int_2 == "agonist" |
                                interactions$int_3 == "agonist" |
                                interactions$int_4 == "agonist" |
                                interactions$int_5 == "agonist"), 1, 0)

interactions$Antibody <-ifelse((interactions$int_1 == "antibody" |
                                interactions$int_2 == "antibody" |
                                interactions$int_3 == "antibody" |
                                interactions$int_4 == "antibody" |
                                interactions$int_5 == "antibody"), 1, 0)

interactions$Activator <-ifelse((interactions$int_1 == "activator" |
                                interactions$int_2 == "activator" |
                                interactions$int_3 == "activator" |
                                interactions$int_4 == "activator" |
                                interactions$int_5 == "activator"), 1, 0)

interactions$Binder <-ifelse((interactions$int_1 == "binder" |
                                interactions$int_2 == "binder" |
                                interactions$int_3 == "binder" |
                                interactions$int_4 == "binder" |
                                interactions$int_5 == "binder"), 1, 0)

interactions$Blocker <-ifelse((interactions$int_1 == "blocker" | 
                                 interactions$int_1 == "channel blocker" |
                                interactions$int_2 == "blocker" | 
                                 interactions$int_2 == "channel blocker" |
                                interactions$int_3 == "blocker" | 
                                 interactions$int_3 == "channel blocker" |
                                interactions$int_4 == "blocker" | 
                                 interactions$int_4 == "channel blocker" |
                                interactions$int_5 == "blocker" | 
                                 interactions$int_5 == "channel blocker"), 1, 0)

interactions$Other <-ifelse((interactions$int_1 == "antisense" |
                               interactions$int_1 == "vaccine" |
                                interactions$int_2 == "antisense" |
                               interactions$int_2 == "vaccine" |
                                interactions$int_3 == "antisense" |
                               interactions$int_3 == "vaccine" |
                                interactions$int_4 == "antisense" |
                               interactions$int_4 == "vaccine" |
                                interactions$int_5 == "antisense" |
                               interactions$int_5 == "vaccine"), 1, 0)


interactions_final <- dplyr::select(interactions, -int_1, -int_2, -int_3, -int_4, -int_5)
interactions_final


# Aggregate and sum by gene name 
interactions_agg <- aggregate(. ~ gene, data = interactions_final, max)

# Order rows to match heatmap
target_int <- as.vector(druggable_dgidb)
interactions_agg <- interactions_agg[match(target_int, interactions_agg$gene),]

# Label row names for heatmap
row.names(interactions_agg) <- target_int
interactions_agg

# Remove gene column
interactions_agg <- dplyr::select(interactions_agg, -gene)

# Replace NA with 0 for plotting
interactions_agg[is.na(interactions_agg)] <- 0

interactions_agg$Inhibitor[interactions_agg$Inhibitor == 1] <- "Inhibitor"
interactions_agg$Antagonist[interactions_agg$Antagonist == 1] <- "Antagonist"
interactions_agg$Agonist[interactions_agg$Agonist == 1] <- "Agonist"
interactions_agg$Antibody[interactions_agg$Antibody == 1] <- "Antibody"
interactions_agg$Activator[interactions_agg$Activator == 1] <- "Activator"
interactions_agg$Binder[interactions_agg$Binder == 1] <- "Binder"
interactions_agg$Blocker[interactions_agg$Blocker == 1] <- "Blocker"
interactions_agg$Other[interactions_agg$Other == 1] <- "Other"

interactions_agg[interactions_agg == 0 ] <- "NA"

interactions_agg

```


```{r phos fda approved drugs}

# FDA approved drugs 
fda <- read.table("./druggable/druggable_FDA_approved_2019-04-17_list.txt")
fda$FDA_approved <- "True"
colnames(fda)[1] <- c("gene")

# Intersect with druggable genes
fda_int <- as.data.frame(row.names(interactions_agg))
colnames(fda_int) <- c("gene")
fda_int <- left_join(fda_int, fda, by = "gene")

# Change NA to 
fda_int$FDA_approved[is.na(fda_int$FDA_approved) == TRUE] <- "False"
fda_int

```


```{r phos heatmap}

### Final Figure Heatmap Phosphoproteome: Druggable Filtered ###`

# Druggable Heatmap
phos_drug_dgidb_annotation <- HeatmapAnnotation(Comparison = c(rep('EGFR', 5), rep('KRAS', 5), 
                                                    rep('TP53', 5), rep('Region', 3), 
                                                    rep('Gender', 3), rep('Smoker', 5)), 
                                       Cohort = c("All", "Female", "Male", "Asian", "Western",
                                                  "All", "Female", "Male", "Asian", "Western", 
                                                  "All", "Female", "Male", "Asian", "Western", 
                                                  "All", "Female", "Male", 
                                                  "All", "Asian", "Western", 
                                                  "All", "Female", "Western", 
                                                  "EGFR_Mutation", "TP53_Mutation"),
                                       col = list(Comparison = c(EGFR = "#EF7C12",
                                                              KRAS = "#FFB90F",
                                                              TP53 = "#F7E690",
                                                              Region = "#EFC7E6",
                                                              Gender = "#CB87B4",
                                                              Smoker = "#B25D91"),
                                                  Cohort = c(All = "Black",
                                                             Female = "#AFDFEF",
                                                             Male = "#54BCD1" ,
                                                              Asian = "#0076C0",
                                                              Western = "#27408B", ##27408B
                                                            EGFR_Mutation = "#EDA9AB",
                                                              TP53_Mutation = "#D9565C")),
                                       annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                                   legend_width = unit(0.20, "cm"),
                                                                   title_gp = gpar(fontsize = 11),
                                                                   labels_gp = gpar(fontsize = 9)),
                                       name = "phos_heat_anno")

phos_heat_drug_dgidb <- Heatmap(phos_dgidb_gene,
                                name = "phos_heat",
                         top_annotation = phos_drug_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 6),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Q Value",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colorRamp2(c(0, 1.0e-50, 0.0001, 0.01), 
                                          c("gray90", "red3", "orangered2", "oldlace")),
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))

# Decorate heatmap
phos_heat_drug_dgidb_fig <- draw(phos_heat_drug_dgidb, heatmap_legend_side = "right")


# Creat heatmap color scheme
colors <-  structure(c("#3C5541", "#3C8C4D", "#3CBC38", 
                       "#AFDE62", "#DFCEE0", "#CDB5CD",
                       "#7473A6", "#132157", "gray90"), 
                     names = c("Antagonist", "Inhibitor", "Agonist", "Antibody", 
                               "Activator", "Binder",
                               "Blocker", "Other", "NA"))

# Heatmap annotations 
#long_cn <- colnames(detail_agg_drug)
#ha_rot_cn <- HeatmapAnnotation(text = anno_text(long_cn, rot = 45, just = "left", offset = unit(1, "mm"), gp = gpar(fontsize = 7)))

ha_rot_cn_dgidb <- HeatmapAnnotation(Interaction = c("Antagonist","Inhibitor","Agonist", "Antibody", 
                                                     "Activator", "Binder", "Blocker", "Other"),
                               col = list(Interaction = c(Antagonist = "#3C5541",
                                                          Inhibitor = "#3C8C4D",
                                                          Agonist = "#3CBC38",
                                                          Antibody = "#AFDE62",
                                                          Activator = "#DFCEE0",
                                                          Binder = "#CDB5CD",
                                                          Blocker = "#7473A6",
                                                          Other = "#132157")),
                               annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                              title = "Interaction",
                                                              legend_width = unit(0.20, "cm"),
                                                              title_gp = gpar(fontsize = 11),
                                                              labels_gp = gpar(fontsize = 9)),
                               show_legend = FALSE)


# Barplot
row_anno_dgidb <- rowAnnotation(barplot = row_anno_barplot(summary_drug_dgidb_score$Log_Score,
                          axis_direction = "reverse", border = TRUE, axis = TRUE, axis_side = "top",
                          gp = gpar(fill = "black")), 
                          width = unit(2, "cm"))

                             
# Create heatmap
interactions_drug_dgidb_heat <- Heatmap(interactions_agg,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Interaction",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors,
                         row_title_gp = gpar(fontsize = 10),
                         column_names_side = "top",
                         top_annotation = ha_rot_cn_dgidb,
                         width = unit(2, "cm"),
                         rect_gp = gpar(col = "white", lwd = 0.75),
                         name = "phos_int")

# Colors for FDA heatmap
colors_2 <-  structure(c("#007BC3", "#AFDFEF"), 
                     names = c("True", "False"))


# Create FDA heatmap
fda_drug_dgidb_heat <- Heatmap(fda_int$FDA_approved,
                         show_column_names = F, 
                         show_row_names = F,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "FDA Approved",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors_2,
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))


#detail_drug_heat
lgd <- Legend(labels = "Score", title = "Druggability", 
              legend_gp = gpar(fill = "black"),
              title_gp = gpar(fontsize = 11),
              labels_gp = gpar(fontsize = 9))

```


```{r phos heatmap decoration}

final_druggable <- draw(row_anno_dgidb + fda_drug_dgidb_heat +
                          interactions_drug_dgidb_heat + phos_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("phos_heat", {
	grid.rect(0, 0, width = 5/26, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(5/26, 0, width = 5/26, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(10/26, 0, width = 5/26, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(15/26, 0, width = 3/26, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(18/26, 0, width = 3/26, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(21/26, 0, width = 5/26, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


```


```{r phos save heatmap}

png("phos_dgidb_final_druggable_heatmap.png", units="in", width=10, height=6, res=800)

final_druggable <- draw(row_anno_dgidb + fda_drug_dgidb_heat +
                          interactions_drug_dgidb_heat + phos_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("phos_heat", {
	grid.rect(0, 0, width = 5/26, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(5/26, 0, width = 5/26, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(10/26, 0, width = 5/26, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(15/26, 0, width = 3/26, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(18/26, 0, width = 3/26, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_heat", {
	grid.rect(21/26, 0, width = 5/26, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("phos_int", {
	grid.rect(0, 0, width = 1, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


dev.off()


```


---


---


---


---

# Proteome FDR 0.05

---


```{r prot read in gene lists and merge}

##### All Upregulated, filtered for DGIDB direct interaction list: Proteome #####

# Read in all proteome significant outliers 
# join all by gene name
# leave significance 
# make this into a heatmap


gender_all_prot <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_up_female__male____sig_outliers.txt", header = TRUE, sep = "\t")
gender_all_prot$GeneSymbol <- as.character(gender_all_prot$GeneSymbol)
colnames(gender_all_prot)[2] <- "gender_all_prot"

region_all_prot <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_up_asian__western____sig_outliers.txt", header = TRUE, sep = "\t")
region_all_prot$GeneSymbol <- as.character(region_all_prot$GeneSymbol)
colnames(region_all_prot)[2] <- "region_all_prot"

#smoking_all_prot <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_up_smoker__non-smoker____sig_outliers.txt", header = TRUE, sep = "\t")
#smoking_all_prot$GeneSymbol <- as.character(smoking_all_prot$GeneSymbol)
#colnames(smoking_all_prot)[2] <- "smoking_all_prot"

egfr_all_prot <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_EGFR_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
egfr_all_prot$GeneSymbol <- as.character(egfr_all_prot$GeneSymbol)
colnames(egfr_all_prot)[2] <- "egfr_all_prot"

kras_all_prot <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_KRAS_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
kras_all_prot$GeneSymbol <- as.character(kras_all_prot$GeneSymbol)
colnames(kras_all_prot)[2] <- "kras_all_prot"

tp53_all_prot <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_TP53_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
tp53_all_prot$GeneSymbol <- as.character(tp53_all_prot$GeneSymbol)
colnames(tp53_all_prot)[2] <- "tp53_all_prot"

#egfr_female_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_EGFR_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#egfr_female_prot$GeneSymbol <- as.character(egfr_female_prot$GeneSymbol)
#colnames(egfr_female_prot)[2] <- "egfr_female_prot"

egfr_male_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_EGFR_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_male_prot$GeneSymbol <- as.character(egfr_male_prot$GeneSymbol)
colnames(egfr_male_prot)[2] <- "egfr_male_prot"

#kras_female_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_KRAS_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#kras_female_prot$GeneSymbol <- as.character(kras_female_prot$GeneSymbol)
#colnames(kras_female_prot)[2] <- "kras_female_prot"

kras_male_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_KRAS_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_male_prot$GeneSymbol <- as.character(kras_male_prot$GeneSymbol)
colnames(kras_male_prot)[2] <- "kras_male_prot"

#tp53_female_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_TP53_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_female_prot$GeneSymbol <- as.character(tp53_female_prot$GeneSymbol)
#colnames(tp53_female_prot)[2] <- "tp53_female_prot"

tp53_male_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Gender_TP53_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_male_prot$GeneSymbol <- as.character(tp53_male_prot$GeneSymbol)
colnames(tp53_male_prot)[2] <- "tp53_male_prot"

egfr_asian_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_EGFR_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_asian_prot$GeneSymbol <- as.character(egfr_asian_prot$GeneSymbol)
colnames(egfr_asian_prot)[2] <- "egfr_asian_prot"

egfr_western_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_EGFR_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_western_prot$GeneSymbol <- as.character(egfr_western_prot$GeneSymbol)
colnames(egfr_western_prot)[2] <- "egfr_western_prot"

gender_asian_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_Gender_up_asian_female_asian_male___sig_outliers.txt", header = TRUE, sep = "\t")
gender_asian_prot$GeneSymbol <- as.character(gender_asian_prot$GeneSymbol)
colnames(gender_asian_prot)[2] <- "gender_asian_prot"

#region_female_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_Gender_up_asian_female_western_female___sig_outliers.txt", header = TRUE, sep = "\t")
#region_female_prot$GeneSymbol <- as.character(region_female_prot$GeneSymbol)
#colnames(region_female_prot)[2] <- "region_female_prot"

region_male_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_Gender_up_asian_male_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
region_male_prot$GeneSymbol <- as.character(region_male_prot$GeneSymbol)
colnames(region_male_prot)[2] <- "region_male_prot"

gender_western_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_Gender_up_western_female_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
gender_western_prot$GeneSymbol <- as.character(gender_western_prot$GeneSymbol)
colnames(gender_western_prot)[2] <- "gender_western_prot"

kras_asian_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_KRAS_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_asian_prot$GeneSymbol <- as.character(kras_asian_prot$GeneSymbol)
colnames(kras_asian_prot)[2] <- "kras_asian_prot"

kras_western_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_KRAS_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_western_prot$GeneSymbol <- as.character(kras_western_prot$GeneSymbol)
colnames(kras_western_prot)[2] <- "kras_western_prot"

#tp53_asian_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_TP53_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_asian_prot$GeneSymbol <- as.character(tp53_asian_prot$GeneSymbol)
#colnames(tp53_asian_prot)[2] <- "tp53_asian_prot"

#tp53_western_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Region_TP53_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_western_prot$GeneSymbol <- as.character(tp53_western_prot$GeneSymbol)
#colnames(tp53_western_prot)[2] <- "tp53_western_prot"

#smoker_female_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_Gender_up_smoker_female_non-smoker_female___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_female_prot$GeneSymbol <- as.character(smoker_female_prot$GeneSymbol)
#colnames(smoker_female_prot)[2] <- "smoker_female_prot"

#smoker_male_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_Gender_up_smoker_male_non-smoker_male___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_male_prot$GeneSymbol <- as.character(smoker_male_prot$GeneSymbol)
#colnames(smoker_male_prot)[2] <- "smoker_male_prot"

#smoker_asian_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_Region_up_smoker_asian_non-smoker_asian___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_asian_prot$GeneSymbol <- as.character(smoker_asian_prot$GeneSymbol)
#colnames(smoker_asian_prot)[2] <- "smoker_asian_prot"

#smoker_western_prot<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_Region_up_smoker_western_non-smoker_western___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_western_prot$GeneSymbol <- as.character(smoker_western_prot$GeneSymbol)
#colnames(smoker_western_prot)[2] <- "smoker_western_prot"

#smoker_egfr_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_EGFR_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_egfr_prot$GeneSymbol <- as.character(smoker_egfr_prot$GeneSymbol)
#colnames(smoker_egfr_prot)[2] <- "smoker_egfr_prot"

#smoker_kras_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_KRAS_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_kras_prot$GeneSymbol <- as.character(smoker_kras_prot$GeneSymbol)
#colnames(smoker_kras_prot)[2] <- "smoker_kras_prot"

#smoker_tp53_prot <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_prot_dgidb_0.05_Smoking_TP53_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_tp53_prot$GeneSymbol <- as.character(smoker_tp53_prot$GeneSymbol)
#colnames(smoker_tp53_prot)[2] <- "smoker_tp53_prot"

###

#prot_dgidb <- full_join(egfr_all_prot, egfr_female_prot, by = "GeneSymbol")
prot_dgidb <- full_join(egfr_all_prot, egfr_male_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, egfr_asian_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, egfr_western_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, kras_all_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, kras_female_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, kras_male_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, kras_asian_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, kras_western_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, tp53_all_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, tp53_female_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, tp53_male_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, tp53_asian_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, tp53_western_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, region_all_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, region_female_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, region_male_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, gender_all_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, gender_asian_prot, by = "GeneSymbol")
prot_dgidb <- full_join(prot_dgidb, gender_western_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoking_all_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_female_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_male_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_asian_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_western_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_egfr_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_tp53_prot, by = "GeneSymbol")
#prot_dgidb <- full_join(prot_dgidb, smoker_kras_prot, by = "GeneSymbol")

prot_dgidb[is.na(prot_dgidb)] <- 0

###

prot_dgidb_gene <- prot_dgidb[,-1]
rownames(prot_dgidb_gene) <- prot_dgidb[,1]
cohorts_all_prot <- colnames(prot_dgidb_gene)
prot_dgidb_annotation <- HeatmapAnnotation(df = as.data.frame(cohorts_all_prot),
                                       col = list(cohorts_all_prot = c(egfr_all_prot = "coral",
                                                              #egfr_female_prot = "dodgerblue4",
                                                              egfr_male_prot = "dodgerblue1",
                                                              egfr_asian_prot= "deepskyblue4",
                                                              egfr_western_prot= "deepskyblue2",
                                                              kras_all_prot = "coral",
                                                              #kras_female_prot = "mediumorchid4",
                                                              kras_male_prot = "mediumorchid1",
                                                              kras_asian_prot= "mediumpurple4",
                                                              kras_western_prot= "mediumpurple1",
                                                              tp53_all_prot = "coral",
                                                              #tp53_female_prot = "olivedrab4",
                                                              tp53_male_prot = "olivedrab3",
                                                              #tp53_asian_prot= "darkgreen",
                                                              #tp53_western_prot= "springgreen4",
                                                              region_all_prot = "coral",
                                                              region_female_prot = "darkorange3",
                                                              region_male_prot = "darkorange",
                                                              gender_all_prot = "coral",
                                                              gender_asian_prot= "darkgoldenrod3",
                                                              gender_western_prot= "darkgoldenrod1")))
                                                              #smoking_all_prot = "coral",
                                                              #smoker_female_prot = "lightsteelblue4",
                                                              #smoker_male_prot = "lightsteelblue1",
                                                              #smoker_asian_prot= "mistyrose4",
                                                              #smoker_western_prot= "mistyrose2",
                                                              #smoker_egfr_prot = "aquamarine4",
                                                              #smoker_kras_prot = "aquamarine3",
                                                              #smoker_tp53_prot = "aquamarine1")))

prot_dgidb_heat <- Heatmap(prot_dgidb_gene,
                         top_annotation = prot_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 5),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(3, "in"),
                                                     title = NULL),
                         col = colorRamp2(c(0, 1.0e-50, 0.005, 0.05), 
                                          c("gray85", "red3", "orangered2", "oldlace")),
                         row_title = "Q Value",
                         row_title_gp = gpar(fontsize = 10))
  
prot_dgidb_heat_2 <- draw(prot_dgidb_heat, heatmap_legend_side = "left")

```


```{r prot dgidb query}

# DGIDB
prot_dgidb_gene

# Filterd
druggable_dgidb_prot <- rownames(prot_dgidb_gene)
length(druggable_dgidb_prot) #63
druggable_dgidb_prot

# Query 
result_drug_dgidb_prot <- queryDGIdb(druggable_dgidb_prot)

result_drug_dgidb_prot

```


```{r prot dgidb result summary}

## Result summary
# Score is the total number of source databases listing the interaction (including PubMed articles)
summary_drug_dgidb_prot <- resultSummary(result_drug_dgidb_prot) 
summary_drug_dgidb_score_prot <- dplyr::select(summary_drug_dgidb_prot, Gene, Score)

# Change to numeric and aggregate for total score
summary_drug_dgidb_score_prot$Score <- as.numeric(summary_drug_dgidb_score_prot$Score)
summary_drug_dgidb_score_prot <- aggregate(. ~ Gene, data = summary_drug_dgidb_score_prot, sum)

# Get log score
summary_drug_dgidb_score_prot$Log_Score <- log(summary_drug_dgidb_score_prot$Score)

# Order rows to match heatmap
target_prot <- as.vector(rownames(prot_dgidb_gene))
target_prot
summary_drug_dgidb_score_prot <- summary_drug_dgidb_score_prot[match(target_prot, summary_drug_dgidb_score_prot$Gene),]

# Label row names for heatmap
row.names(summary_drug_dgidb_score_prot) <- target_prot
summary_drug_dgidb_score_prot

# Replace NA with 0 for plotting
summary_drug_dgidb_score_prot$Score[is.na(summary_drug_dgidb_score_prot$Score) == TRUE] <- 0
summary_drug_dgidb_score_prot$Log_Score[is.na(summary_drug_dgidb_score_prot$Log_Score) == TRUE] <- 0


summary_drug_dgidb_score_prot

```


```{r prot import interactions and aggregate}

# Import manual interactions then aggregate and separate into 1/0
interactions_prot <- read.delim("./druggable/DGIDB_gene_drug_interactions_20190417_modified_dup_removed.txt", sep = "\t")
interactions_prot <- interactions_prot[interactions_prot$gene_name %in% druggable_dgidb_prot ,]

# Select gene name and interaction type only
interactions_prot <- dplyr::select(interactions_prot, gene_name, interaction_types)

# Aggregate to a list
interactions_prot <- aggregate(interaction_types ~ gene_name, interactions_prot, paste, collapse = ",")

# Order rows to match heatmap
target_int_prot <- as.vector(druggable_dgidb_prot)
interactions_prot <- interactions_prot[match(target_int_prot, interactions_prot$gene_name),]

interactions_prot

```


```{r prot manual interactions data frame}

# Rename columns
colnames(interactions_prot)[1:2] <- c("gene", "interaction_type")


# if interaction_type includes this exact match, put 1 else put 0 

interactions_prot <- separate(interactions_prot, interaction_type, 
                         into=c("int_1", "int_2", "int_3", "int_4", "int_5"), sep="," )
interactions_prot$int_1[interactions_prot$int_1==''] <- NA
interactions_prot$int_1[is.na(interactions_prot$int_1) == TRUE] <- 0
interactions_prot$int_2[is.na(interactions_prot$int_2) == TRUE] <- 0
interactions_prot$int_3[is.na(interactions_prot$int_3) == TRUE] <- 0
interactions_prot$int_4[is.na(interactions_prot$int_4) == TRUE] <- 0
interactions_prot$int_5[is.na(interactions_prot$int_5) == TRUE] <- 0

interactions_prot

interactions_prot$Antagonist <-ifelse((interactions_prot$int_1 == "antagonist" |
                                interactions_prot$int_2 == "antagonist" |
                                interactions_prot$int_3 == "antagonist" |
                                interactions_prot$int_4 == "antagonist" |
                                interactions_prot$int_5 == "antagonist"), 1, 0)

interactions_prot$Inhibitor <-ifelse((interactions_prot$int_1 == "inhibitor" |
                                interactions_prot$int_2 == "inhibitor" |
                                interactions_prot$int_3 == "inhibitor" |
                                interactions_prot$int_4 == "inhibitor" |
                                interactions_prot$int_5 == "inhibitor"), 1, 0)

interactions_prot$Agonist <-ifelse((interactions_prot$int_1 == "agonist" |
                                interactions_prot$int_2 == "agonist" |
                                interactions_prot$int_3 == "agonist" |
                                interactions_prot$int_4 == "agonist" |
                                interactions_prot$int_5 == "agonist"), 1, 0)

interactions_prot$Antibody <-ifelse((interactions_prot$int_1 == "antibody" |
                                interactions_prot$int_2 == "antibody" |
                                interactions_prot$int_3 == "antibody" |
                                interactions_prot$int_4 == "antibody" |
                                interactions_prot$int_5 == "antibody"), 1, 0)

interactions_prot$Activator <-ifelse((interactions_prot$int_1 == "activator" |
                                interactions_prot$int_2 == "activator" |
                                interactions_prot$int_3 == "activator" |
                                interactions_prot$int_4 == "activator" |
                                interactions_prot$int_5 == "activator"), 1, 0)

interactions_prot$Binder <-ifelse((interactions_prot$int_1 == "binder" |
                                interactions_prot$int_2 == "binder" |
                                interactions_prot$int_3 == "binder" |
                                interactions_prot$int_4 == "binder" |
                                interactions_prot$int_5 == "binder"), 1, 0)

interactions_prot$Blocker <-ifelse((interactions_prot$int_1 == "blocker" | 
                                 interactions_prot$int_1 == "channel blocker" |
                                interactions_prot$int_2 == "blocker" | 
                                 interactions_prot$int_2 == "channel blocker" |
                                interactions_prot$int_3 == "blocker" | 
                                 interactions_prot$int_3 == "channel blocker" |
                                interactions_prot$int_4 == "blocker" | 
                                 interactions_prot$int_4 == "channel blocker" |
                                interactions_prot$int_5 == "blocker" | 
                                 interactions_prot$int_5 == "channel blocker"), 1, 0)

interactions_prot$Other <-ifelse((interactions_prot$int_1 == "antisense" |
                               interactions_prot$int_1 == "vaccine" | 
                                 interactions_prot$int_1 == "stimulator"| 
                                 interactions_prot$int_1 == "positive allosteric modulator"| 
                                 interactions_prot$int_1 == "cofactor"|
                                interactions_prot$int_2 == "antisense" |
                               interactions_prot$int_2 == "vaccine" |
                                 interactions_prot$int_2 == "stimulator" | 
                                 interactions_prot$int_2 == "positive allosteric modulator"| 
                                 interactions_prot$int_2 == "cofactor"|
                                interactions_prot$int_3 == "antisense" |
                               interactions_prot$int_3 == "vaccine" |
                                 interactions_prot$int_3 == "stimulator" | 
                                 interactions_prot$int_3 == "positive allosteric modulator"| 
                                 interactions_prot$int_3 == "cofactor"|
                                interactions_prot$int_4 == "antisense" |
                               interactions_prot$int_4 == "vaccine" |
                                 interactions_prot$int_4 == "stimulator" | 
                                 interactions_prot$int_4 == "positive allosteric modulator"| 
                                 interactions_prot$int_4 == "cofactor"|
                                interactions_prot$int_5 == "antisense" |
                               interactions_prot$int_5 == "vaccine" |
                                 interactions_prot$int_5 == "stimulator"| 
                                 interactions_prot$int_5 == "positive allosteric modulator" | 
                                 interactions_prot$int_5 == "cofactor"), 1, 0)


interactions_prot_final <- dplyr::select(interactions_prot, -int_1, -int_2, -int_3, -int_4, -int_5)
interactions_prot_final


# Aggregate and sum by gene name 
interactions_prot_agg <- aggregate(. ~ gene, data = interactions_prot_final, max)

# Order rows to match heatmap
target_int_prot <- as.vector(druggable_dgidb_prot)
interactions_prot_agg <- interactions_prot_agg[match(target_int_prot, interactions_prot_agg$gene),]

# Label row names for heatmap
row.names(interactions_prot_agg) <- target_int_prot
interactions_prot_agg

# Remove gene column
interactions_prot_agg <- dplyr::select(interactions_prot_agg, -gene)

# Replace NA with 0 for plotting
interactions_prot_agg[is.na(interactions_prot_agg)] <- 0

interactions_prot_agg$Inhibitor[interactions_prot_agg$Inhibitor == 1] <- "Inhibitor"
interactions_prot_agg$Antagonist[interactions_prot_agg$Antagonist == 1] <- "Antagonist"
interactions_prot_agg$Agonist[interactions_prot_agg$Agonist == 1] <- "Agonist"
interactions_prot_agg$Antibody[interactions_prot_agg$Antibody == 1] <- "Antibody"
interactions_prot_agg$Activator[interactions_prot_agg$Activator == 1] <- "Activator"
interactions_prot_agg$Binder[interactions_prot_agg$Binder == 1] <- "Binder"
interactions_prot_agg$Blocker[interactions_prot_agg$Blocker == 1] <- "Blocker"
interactions_prot_agg$Other[interactions_prot_agg$Other == 1] <- "Other"

interactions_prot_agg[interactions_prot_agg == 0 ] <- "NA"

interactions_prot_agg

```


```{r prot fda approved drugs}

# FDA approved drugs 
fda_prot <- read.table("./druggable/druggable_FDA_approved_2019-04-17_list.txt")
fda_prot$FDA_approved <- "True"
colnames(fda_prot)[1] <- c("gene")

# Intersect with druggable genes
fda_int_prot <- as.data.frame(row.names(interactions_prot_agg))
colnames(fda_int_prot) <- c("gene")
fda_int_prot <- left_join(fda_int_prot, fda_prot, by = "gene")

# Change NA to 
fda_int_prot$FDA_approved[is.na(fda_int_prot$FDA_approved) == TRUE] <- "False"
fda_int_prot

```


```{r prot heatmap}

### Final Figure Heatmap: Druggable Filtered ###`

# Druggable Heatmap
prot_drug_dgidb_annotation <- HeatmapAnnotation(Comparison = c(rep('EGFR', 4), rep('KRAS', 4), 
                                                    rep('TP53', 2), rep('Region', 2), 
                                                    rep('Gender', 3)), 
                                       Cohort = c("All", "Male", "Asian", "Western",
                                                  "All", "Male", "Asian", "Western", 
                                                  "All", "Male",  
                                                  "All", "Male", 
                                                  "All", "Asian", "Western"),
                                       col = list(Comparison = c(EGFR = "#EF7C12",
                                                              KRAS = "#FFB90F",
                                                              TP53 = "#F7E690",
                                                              Region = "#EFC7E6",
                                                              Gender = "#CB87B4"),
                                                              #Smoker = "#B25D91"),
                                                  Cohort = c(All = "Black",
                                                             #Female = "#AFDFEF",
                                                             Male = "#54BCD1" ,
                                                              Asian = "#0076C0",
                                                              Western = "#27408B")),
                                                            #EGFR_Mutation = "#EDA9AB",
                                                              #TP53_Mutation = "#D9565C")),
                                       annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                                   legend_width = unit(0.20, "cm"),
                                                                   title_gp = gpar(fontsize = 11),
                                                                   labels_gp = gpar(fontsize = 9)),
                                       name = "prot_heat_anno")

prot_heat_drug_dgidb <- Heatmap(prot_dgidb_gene,
                                name = "prot_heat",
                         top_annotation = prot_drug_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 7),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Q Value",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colorRamp2(c(0, 1.0e-50, 0.01, 0.05), 
                                          c("gray90", "red3", "orangered2", "oldlace")),
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))

# Decorate heatmap
prot_heat_drug_dgidb_fig <- draw(prot_heat_drug_dgidb, heatmap_legend_side = "right")


# Creat heatmap color scheme
colors <-  structure(c("#3C5541", "#3C8C4D", "#3CBC38", 
                       "#AFDE62", "#DFCEE0", "#CDB5CD",
                       "#7473A6", "#132157", "gray90"), 
                     names = c("Antagonist", "Inhibitor", "Agonist", "Antibody", 
                               "Activator", "Binder",
                               "Blocker", "Other", "NA"))

# Heatmap annotations 
ha_rot_cn_dgidb <- HeatmapAnnotation(Interaction = c("Antagonist","Inhibitor","Agonist", "Antibody", 
                                                     "Activator", "Binder", "Blocker", "Other"),
                               col = list(Interaction = c(Antagonist = "#3C5541",
                                                          Inhibitor = "#3C8C4D",
                                                          Agonist = "#3CBC38",
                                                          Antibody = "#AFDE62",
                                                          Activator = "#DFCEE0",
                                                          Binder = "#CDB5CD",
                                                          Blocker = "#7473A6",
                                                          Other = "#132157")),
                               annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                              title = "Interaction",
                                                              legend_width = unit(0.20, "cm"),
                                                              title_gp = gpar(fontsize = 11),
                                                              labels_gp = gpar(fontsize = 9)),
                               show_legend = FALSE)


# Barplot
row_anno_dgidb_prot <- rowAnnotation(barplot = row_anno_barplot(summary_drug_dgidb_score_prot$Log_Score,
                          axis_direction = "reverse", border = TRUE, axis = TRUE, axis_side = "top",
                          gp = gpar(fill = "black")), 
                          width = unit(2, "cm"))

                             
# Create heatmap
interactions_prot_drug_dgidb_heat <- Heatmap(interactions_prot_agg,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Interaction",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors,
                         row_title_gp = gpar(fontsize = 10),
                         column_names_side = "top",
                         top_annotation = ha_rot_cn_dgidb,
                         width = unit(2, "cm"),
                         rect_gp = gpar(col = "white", lwd = 0.75),
                         name = "prot_int")

# Colors for FDA heatmap
colors_2 <-  structure(c("#007BC3", "#AFDFEF"), 
                     names = c("True", "False"))


# Create FDA heatmap
fda_drug_dgidb_heat_prot <- Heatmap(fda_int_prot$FDA_approved,
                         show_column_names = F, 
                         show_row_names = F,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "FDA Approved",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors_2,
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))


#detail_drug_heat
lgd <- Legend(labels = "Score", title = "Druggability", 
              legend_gp = gpar(fill = "black"),
              title_gp = gpar(fontsize = 11),
              labels_gp = gpar(fontsize = 9))

```


```{r prot heatmap decorations}

final_druggable_prot <- draw(row_anno_dgidb_prot + fda_drug_dgidb_heat_prot +
                          interactions_prot_drug_dgidb_heat + prot_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("prot_heat", {
	grid.rect(0, 0, width = 4/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(4/15, 0, width = 4/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(8/15, 0, width = 2/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(10/15, 0, width = 2/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(12/15, 0, width = 3/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


decorate_heatmap_body("prot_int", {
	grid.rect(0, 0, width = 1, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


```


```{r prot heatmap save}

png("prot_dgidb_final_druggable_heatmap.png", units="in", width=9, height=4.5, res=800)

final_druggable_prot <- draw(row_anno_dgidb_prot + fda_drug_dgidb_heat_prot +
                          interactions_prot_drug_dgidb_heat + prot_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("prot_heat", {
	grid.rect(0, 0, width = 4/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(4/15, 0, width = 4/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(8/15, 0, width = 2/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(10/15, 0, width = 2/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("prot_heat", {
	grid.rect(12/15, 0, width = 3/15, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


decorate_heatmap_body("prot_int", {
	grid.rect(0, 0, width = 1, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

dev.off()

```


---


---


---


---

# RNA Log 2 FDR 0.05

---


```{r rna read in gene lists and merge}

##### All Upregulated, filtered for DGIDB direct interaction list: RNA Log 2 #####

# Read in all rna significant outliers 
# join all by gene name
# leave significance 
# make this into a heatmap


gender_all_rna <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_up_female__male____sig_outliers.txt", header = TRUE, sep = "\t")
gender_all_rna$GeneSymbol <- as.character(gender_all_rna$GeneSymbol)
colnames(gender_all_rna)[2] <- "gender_all_rna"

region_all_rna <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_up_asian__western____sig_outliers.txt", header = TRUE, sep = "\t")
region_all_rna$GeneSymbol <- as.character(region_all_rna$GeneSymbol)
colnames(region_all_rna)[2] <- "region_all_rna"

#smoking_all_rna <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_up_smoker__non-smoker____sig_outliers.txt", header = TRUE, sep = "\t")
#smoking_all_rna$GeneSymbol <- as.character(smoking_all_rna$GeneSymbol)
#colnames(smoking_all_rna)[2] <- "smoking_all_rna"

egfr_all_rna <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_EGFR_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
egfr_all_rna$GeneSymbol <- as.character(egfr_all_rna$GeneSymbol)
colnames(egfr_all_rna)[2] <- "egfr_all_rna"

kras_all_rna <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_KRAS_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
kras_all_rna$GeneSymbol <- as.character(kras_all_rna$GeneSymbol)
colnames(kras_all_rna)[2] <- "kras_all_rna"

tp53_all_rna <- read.table("./comparison_results_1_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_TP53_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
tp53_all_rna$GeneSymbol <- as.character(tp53_all_rna$GeneSymbol)
colnames(tp53_all_rna)[2] <- "tp53_all_rna"

#egfr_female_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_EGFR_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#egfr_female_rna$GeneSymbol <- as.character(egfr_female_rna$GeneSymbol)
#colnames(egfr_female_rna)[2] <- "egfr_female_rna"

egfr_male_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_EGFR_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_male_rna$GeneSymbol <- as.character(egfr_male_rna$GeneSymbol)
colnames(egfr_male_rna)[2] <- "egfr_male_rna"

#kras_female_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_KRAS_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#kras_female_rna$GeneSymbol <- as.character(kras_female_rna$GeneSymbol)
#colnames(kras_female_rna)[2] <- "kras_female_rna"

kras_male_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_KRAS_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_male_rna$GeneSymbol <- as.character(kras_male_rna$GeneSymbol)
colnames(kras_male_rna)[2] <- "kras_male_rna"

#tp53_female_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_TP53_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_female_rna$GeneSymbol <- as.character(tp53_female_rna$GeneSymbol)
#colnames(tp53_female_rna)[2] <- "tp53_female_rna"

tp53_male_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Gender_TP53_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_male_rna$GeneSymbol <- as.character(tp53_male_rna$GeneSymbol)
colnames(tp53_male_rna)[2] <- "tp53_male_rna"

egfr_asian_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_EGFR_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_asian_rna$GeneSymbol <- as.character(egfr_asian_rna$GeneSymbol)
colnames(egfr_asian_rna)[2] <- "egfr_asian_rna"

egfr_western_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_EGFR_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_western_rna$GeneSymbol <- as.character(egfr_western_rna$GeneSymbol)
colnames(egfr_western_rna)[2] <- "egfr_western_rna"

gender_asian_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_Gender_up_asian_female_asian_male___sig_outliers.txt", header = TRUE, sep = "\t")
gender_asian_rna$GeneSymbol <- as.character(gender_asian_rna$GeneSymbol)
colnames(gender_asian_rna)[2] <- "gender_asian_rna"

#region_female_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_Gender_up_asian_female_western_female___sig_outliers.txt", header = TRUE, sep = "\t")
#region_female_rna$GeneSymbol <- as.character(region_female_rna$GeneSymbol)
#colnames(region_female_rna)[2] <- "region_female_rna"

region_male_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_Gender_up_asian_male_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
region_male_rna$GeneSymbol <- as.character(region_male_rna$GeneSymbol)
colnames(region_male_rna)[2] <- "region_male_rna"

gender_western_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_Gender_up_western_female_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
gender_western_rna$GeneSymbol <- as.character(gender_western_rna$GeneSymbol)
colnames(gender_western_rna)[2] <- "gender_western_rna"

kras_asian_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_KRAS_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_asian_rna$GeneSymbol <- as.character(kras_asian_rna$GeneSymbol)
colnames(kras_asian_rna)[2] <- "kras_asian_rna"

kras_western_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_KRAS_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
kras_western_rna$GeneSymbol <- as.character(kras_western_rna$GeneSymbol)
colnames(kras_western_rna)[2] <- "kras_western_rna"

tp53_asian_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_TP53_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_asian_rna$GeneSymbol <- as.character(tp53_asian_rna$GeneSymbol)
colnames(tp53_asian_rna)[2] <- "tp53_asian_rna"

tp53_western_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Region_TP53_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_western_rna$GeneSymbol <- as.character(tp53_western_rna$GeneSymbol)
colnames(tp53_western_rna)[2] <- "tp53_western_rna"

smoker_female_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_Gender_up_smoker_female_non-smoker_female___sig_outliers.txt", header = TRUE, sep = "\t")
smoker_female_rna$GeneSymbol <- as.character(smoker_female_rna$GeneSymbol)
colnames(smoker_female_rna)[2] <- "smoker_female_rna"

#smoker_male_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_Gender_up_smoker_male_non-smoker_male___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_male_rna$GeneSymbol <- as.character(smoker_male_rna$GeneSymbol)
#colnames(smoker_male_rna)[2] <- "smoker_male_rna"

#smoker_asian_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_Region_up_smoker_asian_non-smoker_asian___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_asian_rna$GeneSymbol <- as.character(smoker_asian_rna$GeneSymbol)
#colnames(smoker_asian_rna)[2] <- "smoker_asian_rna"

#smoker_western_rna<- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_Region_up_smoker_western_non-smoker_western___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_western_rna$GeneSymbol <- as.character(smoker_western_rna$GeneSymbol)
#colnames(smoker_western_rna)[2] <- "smoker_western_rna"

#smoker_egfr_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_EGFR_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_egfr_rna$GeneSymbol <- as.character(smoker_egfr_rna$GeneSymbol)
#colnames(smoker_egfr_rna)[2] <- "smoker_egfr_rna"

#smoker_kras_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_KRAS_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_kras_rna$GeneSymbol <- as.character(smoker_kras_rna$GeneSymbol)
#colnames(smoker_kras_rna)[2] <- "smoker_kras_rna"

#smoker_tp53_rna <- read.table("./comparison_results_2_dgidb_fdr_0.05/luad2_rna_log2_dgidb_0.05_Smoking_TP53_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_tp53_rna$GeneSymbol <- as.character(smoker_tp53_rna$GeneSymbol)
#colnames(smoker_tp53_rna)[2] <- "smoker_tp53_rna"

###

#rna_dgidb <- full_join(egfr_all_rna, egfr_female_rna, by = "GeneSymbol")
rna_dgidb <- full_join(egfr_all_rna, egfr_male_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, egfr_asian_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, egfr_western_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, kras_all_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, kras_female_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, kras_male_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, kras_asian_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, kras_western_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, tp53_all_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, tp53_female_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, tp53_male_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, tp53_asian_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, tp53_western_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, region_all_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, region_female_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, region_male_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, gender_all_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, gender_asian_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, gender_western_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoking_all_rna, by = "GeneSymbol")
rna_dgidb <- full_join(rna_dgidb, smoker_female_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoker_male_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoker_asian_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoker_western_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoker_egfr_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoker_tp53_rna, by = "GeneSymbol")
#rna_dgidb <- full_join(rna_dgidb, smoker_kras_rna, by = "GeneSymbol")

rna_dgidb[is.na(rna_dgidb)] <- 0

##

rna_dgidb_gene <- rna_dgidb[,-1]
rownames(rna_dgidb_gene) <- rna_dgidb[,1]
cohorts_all_rna <- colnames(rna_dgidb_gene)
rna_dgidb_annotation <- HeatmapAnnotation(df = as.data.frame(cohorts_all_rna),
                                       col = list(cohorts_all_rna = c(egfr_all_rna = "coral",
                                                              #egfr_female_rna = "dodgerblue4",
                                                              egfr_male_rna = "dodgerblue1",
                                                              egfr_asian_rna= "deepskyblue4",
                                                              egfr_western_rna= "deepskyblue2",
                                                              kras_all_rna = "coral",
                                                              #kras_female_rna = "mediumorchid4",
                                                              kras_male_rna = "mediumorchid1",
                                                              kras_asian_rna= "mediumpurple4",
                                                              kras_western_rna= "mediumpurple1",
                                                              tp53_all_rna = "coral",
                                                              #tp53_female_rna = "olivedrab4",
                                                              tp53_male_rna = "olivedrab3",
                                                              tp53_asian_rna= "darkgreen",
                                                              tp53_western_rna= "springgreen4",
                                                              region_all_rna = "coral",
                                                              #region_female_rna = "darkorange3",
                                                              region_male_rna = "darkorange",
                                                              gender_all_rna = "coral",
                                                              gender_asian_rna= "darkgoldenrod3",
                                                              gender_western_rna= "darkgoldenrod1",
                                                              #smoking_all_rna = "coral",
                                                              smoker_female_rna = "lightsteelblue4")))
                                                              #smoker_male_rna = "lightsteelblue1",
                                                              #smoker_asian_rna= "mistyrose4",
                                                              #smoker_western_rna= "mistyrose2",
                                                              #smoker_egfr_rna = "aquamarine4",
                                                              #smoker_kras_rna = "aquamarine3",
                                                              #smoker_tp53_rna = "aquamarine1")))

rna_dgidb_heat <- Heatmap(rna_dgidb_gene,
                         top_annotation = rna_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 5),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(3, "in"),
                                                     title = NULL),
                         col = colorRamp2(c(0, 1.0e-50, 0.005, 0.05), 
                                          c("gray85", "red3", "orangered2", "oldlace")),
                         row_title = "Q Value",
                         row_title_gp = gpar(fontsize = 10))
  
rna_dgidb_heat_2 <- draw(rna_dgidb_heat, heatmap_legend_side = "left")

```


```{r rna dgidb query}

# DGIDB
rna_dgidb_gene

# Filterd
druggable_dgidb_rna <- rownames(rna_dgidb_gene)
length(druggable_dgidb_rna) #63
druggable_dgidb_rna

# Query 
result_drug_dgidb_rna <- queryDGIdb(druggable_dgidb_rna)

result_drug_dgidb_rna

```


```{r rna dgidb result summary}

## Result summary
# Score is the total number of source databases listing the interaction (including PubMed articles)
summary_drug_dgidb_rna <- resultSummary(result_drug_dgidb_rna) 
summary_drug_dgidb_score_rna <- dplyr::select(summary_drug_dgidb_rna, Gene, Score)

# Change to numeric and aggregate for total score
summary_drug_dgidb_score_rna$Score <- as.numeric(summary_drug_dgidb_score_rna$Score)
summary_drug_dgidb_score_rna <- aggregate(. ~ Gene, data = summary_drug_dgidb_score_rna, sum)

# Get log score
summary_drug_dgidb_score_rna$Log_Score <- log(summary_drug_dgidb_score_rna$Score)

# Order rows to match heatmap
target_rna <- as.vector(rownames(rna_dgidb_gene))
target_rna
summary_drug_dgidb_score_rna <- summary_drug_dgidb_score_rna[match(target_rna, summary_drug_dgidb_score_rna$Gene),]

# Label row names for heatmap
row.names(summary_drug_dgidb_score_rna) <- target_rna
summary_drug_dgidb_score_rna

# Replace NA with 0 for plotting
summary_drug_dgidb_score_rna$Score[is.na(summary_drug_dgidb_score_rna$Score) == TRUE] <- 0
summary_drug_dgidb_score_rna$Log_Score[is.na(summary_drug_dgidb_score_rna$Log_Score) == TRUE] <- 0


summary_drug_dgidb_score_rna

```


```{r rna import interactions and aggregate}

# Import manual interactions then aggregate and separate into 1/0
interactions_rna <- read.delim("./druggable/DGIDB_gene_drug_interactions_20190417_modified_dup_removed.txt", sep = "\t")
interactions_rna <- interactions_rna[interactions_rna$gene_name %in% druggable_dgidb_rna ,]

# Select gene name and interaction type only
interactions_rna <- dplyr::select(interactions_rna, gene_name, interaction_types)

# Aggregate to a list
interactions_rna <- aggregate(interaction_types ~ gene_name, interactions_rna, paste, collapse = ",")

# Order rows to match heatmap
target_int_rna <- as.vector(druggable_dgidb_rna)
interactions_rna <- interactions_rna[match(target_int_rna, interactions_rna$gene_name),]

interactions_rna

```


```{r rna manual interactions data frame}

# Rename columns
colnames(interactions_rna)[1:2] <- c("gene", "interaction_type")


# if interaction_type includes this exact match, put 1 else put 0 

interactions_rna <- separate(interactions_rna, interaction_type, 
                         into=c("int_1", "int_2", "int_3", "int_4", "int_5", "int_6"), sep="," )
interactions_rna$int_1[interactions_rna$int_1==''] <- NA
interactions_rna$int_1[is.na(interactions_rna$int_1) == TRUE] <- 0
interactions_rna$int_2[is.na(interactions_rna$int_2) == TRUE] <- 0
interactions_rna$int_3[is.na(interactions_rna$int_3) == TRUE] <- 0
interactions_rna$int_4[is.na(interactions_rna$int_4) == TRUE] <- 0
interactions_rna$int_5[is.na(interactions_rna$int_5) == TRUE] <- 0
interactions_rna$int_6[is.na(interactions_rna$int_6) == TRUE] <- 0

interactions_rna

interactions_rna$Antagonist <-ifelse((interactions_rna$int_1 == "antagonist" |
                                interactions_rna$int_2 == "antagonist" |
                                interactions_rna$int_3 == "antagonist" |
                                interactions_rna$int_4 == "antagonist" |
                                interactions_rna$int_5 == "antagonist" |
                                interactions_rna$int_6 == "antagonist"), 1, 0)

interactions_rna$Inhibitor <-ifelse((interactions_rna$int_1 == "inhibitor" |
                                interactions_rna$int_1 == "gating inhibitor" |
                                interactions_rna$int_2 == "inhibitor" |
                                interactions_rna$int_2 == "gating inhibitor" |
                                interactions_rna$int_3 == "inhibitor" |
                                interactions_rna$int_3 == "gating inhibitor" |
                                interactions_rna$int_4 == "inhibitor" |
                                interactions_rna$int_4 == "gating inhibitor" |
                                interactions_rna$int_5 == "inhibitor"  |
                                interactions_rna$int_5 == "gating inhibitor"|
                                interactions_rna$int_6 == "inhibitor" |
                                interactions_rna$int_6 == "gating inhibitor"), 1, 0)

interactions_rna$Agonist <-ifelse((interactions_rna$int_1 == "agonist"|
                                interactions_rna$int_1 == "partial agonist" |
                                interactions_rna$int_2 == "agonist"|
                                interactions_rna$int_2 == "partial agonist" |
                                interactions_rna$int_3 == "agonist"|
                                interactions_rna$int_3 == "partial agonist" |
                                interactions_rna$int_4 == "agonist"|
                                interactions_rna$int_4 == "partial agonist" |
                                interactions_rna$int_5 == "agonist"|
                                interactions_rna$int_5 == "partial agonist" |
                                interactions_rna$int_6 == "agonist" |
                                interactions_rna$int_6 == "partial agonist"), 1, 0)

interactions_rna$Antibody <-ifelse((interactions_rna$int_1 == "antibody" |
                                interactions_rna$int_2 == "antibody" |
                                interactions_rna$int_3 == "antibody" |
                                interactions_rna$int_4 == "antibody" |
                                interactions_rna$int_5 == "antibody" |
                                interactions_rna$int_6 == "antibody"), 1, 0)

interactions_rna$Activator <-ifelse((interactions_rna$int_1 == "activator" |
                                interactions_rna$int_2 == "activator" |
                                interactions_rna$int_3 == "activator" |
                                interactions_rna$int_4 == "activator" |
                                interactions_rna$int_5 == "activator" |
                                interactions_rna$int_6 == "activator"), 1, 0)

interactions_rna$Binder <-ifelse((interactions_rna$int_1 == "binder" |
                                interactions_rna$int_2 == "binder" |
                                interactions_rna$int_3 == "binder" |
                                interactions_rna$int_4 == "binder" |
                                interactions_rna$int_5 == "binder" |
                                interactions_rna$int_6 == "binder"), 1, 0)

interactions_rna$Blocker <-ifelse((interactions_rna$int_1 == "blocker" | 
                                 interactions_rna$int_1 == "channel blocker" |
                                interactions_rna$int_2 == "blocker" | 
                                 interactions_rna$int_2 == "channel blocker" |
                                interactions_rna$int_3 == "blocker" | 
                                 interactions_rna$int_3 == "channel blocker" |
                                interactions_rna$int_4 == "blocker" | 
                                 interactions_rna$int_4 == "channel blocker" |
                                interactions_rna$int_5 == "blocker" | 
                                 interactions_rna$int_5 == "channel blocker" |
                                interactions_rna$int_6 == "blocker" | 
                                 interactions_rna$int_6 == "channel blocker"), 1, 0)

#allosteric modulator

other_list <- c("antisense", "vaccine", "stimulator", "positive allosteric modulator",
                "allosteric modulator", "cofactor", "antisense oligonucleotide")

interactions_rna$Other <- ifelse(((interactions_rna$int_1 %in% other_list) == TRUE) |
  ((interactions_rna$int_2 %in% other_list) == TRUE) |
  ((interactions_rna$int_3 %in% other_list) == TRUE) |
  ((interactions_rna$int_4 %in% other_list) == TRUE) |
  ((interactions_rna$int_5 %in% other_list) == TRUE) |
  ((interactions_rna$int_6 %in% other_list) == TRUE), 1, 0)


interactions_rna_final <- dplyr::select(interactions_rna, -int_1, -int_2, -int_3, -int_4, -int_5, -int_6)
interactions_rna_final


# Aggregate and sum by gene name 
interactions_rna_agg <- aggregate(. ~ gene, data = interactions_rna_final, max)

# Order rows to match heatmap
target_int_rna <- as.vector(druggable_dgidb_rna)
interactions_rna_agg <- interactions_rna_agg[match(target_int_rna, interactions_rna_agg$gene),]

# Label row names for heatmap
row.names(interactions_rna_agg) <- target_int_rna
interactions_rna_agg

# Remove gene column
interactions_rna_agg <- dplyr::select(interactions_rna_agg, -gene)

# Replace NA with 0 for plotting
interactions_rna_agg[is.na(interactions_rna_agg)] <- 0

interactions_rna_agg$Inhibitor[interactions_rna_agg$Inhibitor == 1] <- "Inhibitor"
interactions_rna_agg$Antagonist[interactions_rna_agg$Antagonist == 1] <- "Antagonist"
interactions_rna_agg$Agonist[interactions_rna_agg$Agonist == 1] <- "Agonist"
interactions_rna_agg$Antibody[interactions_rna_agg$Antibody == 1] <- "Antibody"
interactions_rna_agg$Activator[interactions_rna_agg$Activator == 1] <- "Activator"
interactions_rna_agg$Binder[interactions_rna_agg$Binder == 1] <- "Binder"
interactions_rna_agg$Blocker[interactions_rna_agg$Blocker == 1] <- "Blocker"
interactions_rna_agg$Other[interactions_rna_agg$Other == 1] <- "Other"

interactions_rna_agg[interactions_rna_agg == 0 ] <- "NA"

interactions_rna_agg

```


```{r rna fda approved drugs}

# FDA approved drugs 
fda_rna <- read.table("./druggable/druggable_FDA_approved_2019-04-17_list.txt")
fda_rna$FDA_approved <- "True"
colnames(fda_rna)[1] <- c("gene")

# Intersect with druggable genes
fda_int_rna <- as.data.frame(row.names(interactions_rna_agg))
colnames(fda_int_rna) <- c("gene")
fda_int_rna <- left_join(fda_int_rna, fda_rna, by = "gene")

# Change NA to 
fda_int_rna$FDA_approved[is.na(fda_int_rna$FDA_approved) == TRUE] <- "False"
fda_int_rna

```


```{r rna heatmap}

### Final Figure Heatmap: Druggable Filtered ###`

# Druggable Heatmap
rna_drug_dgidb_annotation <- HeatmapAnnotation(Comparison = c(rep('EGFR', 4), rep('KRAS', 4), 
                                                    rep('TP53', 4), rep('Region', 2), 
                                                    rep('Gender', 3), rep("Smoker", 1)), 
                                       Cohort = c("All", "Male", "Asian", "Western",
                                                  "All", "Male", "Asian", "Western", 
                                                  "All", "Male", "Asian", "Western",  
                                                  "All", "Male", 
                                                  "All", "Asian", "Western",
                                                  "Female"),
                                       col = list(Comparison = c(EGFR = "#EF7C12",
                                                              KRAS = "#FFB90F",
                                                              TP53 = "#F7E690",
                                                              Region = "#EFC7E6",
                                                              Gender = "#CB87B4",
                                                              Smoker = "#B25D91"),
                                                  Cohort = c(All = "Black",
                                                             Female = "#AFDFEF",
                                                             Male = "#54BCD1" ,
                                                              Asian = "#0076C0",
                                                              Western = "#27408B")),
                                                            #EGFR_Mutation = "#EDA9AB",
                                                              #TP53_Mutation = "#D9565C")),
                                       annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                                   legend_width = unit(0.20, "cm"),
                                                                   title_gp = gpar(fontsize = 11),
                                                                   labels_gp = gpar(fontsize = 9)),
                                       name = "rna_heat_anno")

rna_heat_drug_dgidb <- Heatmap(rna_dgidb_gene,
                                name = "rna_heat",
                         top_annotation = rna_drug_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 6),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Q Value",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colorRamp2(c(0, 1.0e-50, 0.01, 0.05), 
                                          c("gray90", "red3", "orangered2", "oldlace")),
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))

# Decorate heatmap
rna_heat_drug_dgidb_fig <- draw(rna_heat_drug_dgidb, heatmap_legend_side = "right")


# Creat heatmap color scheme
colors <-  structure(c("#3C5541", "#3C8C4D", "#3CBC38", 
                       "#AFDE62", "#DFCEE0", "#CDB5CD",
                       "#7473A6", "#132157", "gray90"), 
                     names = c("Antagonist", "Inhibitor", "Agonist", "Antibody", 
                               "Activator", "Binder",
                               "Blocker", "Other", "NA"))

# Heatmap annotations 
ha_rot_cn_dgidb <- HeatmapAnnotation(Interaction = c("Antagonist","Inhibitor","Agonist", "Antibody", 
                                                     "Activator", "Binder", "Blocker", "Other"),
                               col = list(Interaction = c(Antagonist = "#3C5541",
                                                          Inhibitor = "#3C8C4D",
                                                          Agonist = "#3CBC38",
                                                          Antibody = "#AFDE62",
                                                          Activator = "#DFCEE0",
                                                          Binder = "#CDB5CD",
                                                          Blocker = "#7473A6",
                                                          Other = "#132157")),
                               annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                              title = "Interaction",
                                                              legend_width = unit(0.20, "cm"),
                                                              title_gp = gpar(fontsize = 11),
                                                              labels_gp = gpar(fontsize = 9)),
                               show_legend = FALSE)


# Barplot
row_anno_dgidb_rna <- rowAnnotation(barplot = row_anno_barplot(summary_drug_dgidb_score_rna$Log_Score,
                          axis_direction = "reverse", border = TRUE, axis = TRUE, axis_side = "top",
                          gp = gpar(fill = "black")), 
                          width = unit(2, "cm"))

                             
# Create heatmap
interactions_rna_drug_dgidb_heat <- Heatmap(interactions_rna_agg,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Interaction",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors,
                         row_title_gp = gpar(fontsize = 10),
                         column_names_side = "top",
                         top_annotation = ha_rot_cn_dgidb,
                         width = unit(2, "cm"),
                         rect_gp = gpar(col = "white", lwd = 0.75),
                         name = "rna_int")

# Colors for FDA heatmap
colors_2 <-  structure(c("#007BC3", "#AFDFEF"), 
                     names = c("True", "False"))


# Create FDA heatmap
fda_drug_dgidb_heat_rna <- Heatmap(fda_int_rna$FDA_approved,
                         show_column_names = F, 
                         show_row_names = F,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "FDA Approved",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors_2,
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))


#detail_drug_heat
lgd <- Legend(labels = "Score", title = "Druggability", 
              legend_gp = gpar(fill = "black"),
              title_gp = gpar(fontsize = 11),
              labels_gp = gpar(fontsize = 9))

```


```{r rna heatmap decorations}

final_druggable_rna <- draw(row_anno_dgidb_rna + fda_drug_dgidb_heat_rna +
                          interactions_rna_drug_dgidb_heat + rna_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("rna_heat", {
	grid.rect(0, 0, width = 4/18, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(4/18, 0, width = 4/18, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(8/18, 0, width = 4/18, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(12/18, 0, width = 2/18, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(14/18, 0, width = 3/18, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(17/18, 0, width = 1/18, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_int", {
	grid.rect(0, 0, width = 1, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


```


```{r rna heatmap save}

png("rna_dgidb_final_druggable_heatmap.png", units="in", width=10, height=6, res=800)

final_druggable_rna <- draw(row_anno_dgidb_rna + fda_drug_dgidb_heat_rna +
                          interactions_rna_drug_dgidb_heat + rna_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("rna_heat", {
	grid.rect(0, 0, width = 4/18, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(4/18, 0, width = 4/18, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(8/18, 0, width = 4/18, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(12/18, 0, width = 2/18, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(14/18, 0, width = 3/18, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_heat", {
	grid.rect(17/18, 0, width = 1/18, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("rna_int", {
	grid.rect(0, 0, width = 1, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


dev.off()

```


---


---

# CNV FDR 0.01, IQR 1.25

---


```{r cnv read in gene lists and merge}

##### All Upregulated, filtered for DGIDB direct interaction list: CNV  #####

# Read in all cnv significant outliers 
# join all by gene name
# leave significance 
# make this into a heatmap


#gender_all_cnv <- read.table("./comparison_results_1_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_up_female__male____sig_outliers.txt", header = TRUE, sep = "\t")
#gender_all_cnv$GeneSymbol <- as.character(gender_all_cnv$GeneSymbol)
#colnames(gender_all_cnv)[2] <- "gender_all_cnv"

#region_all_cnv <- read.table("./comparison_results_1_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_up_asian__western____sig_outliers.txt", header = TRUE, sep = "\t")
#region_all_cnv$GeneSymbol <- as.character(region_all_cnv$GeneSymbol)
#colnames(region_all_cnv)[2] <- "region_all_cnv"

#smoking_all_cnv <- read.table("./comparison_results_1_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_up_smoker__non-smoker____sig_outliers.txt", header = TRUE, sep = "\t")
#smoking_all_cnv$GeneSymbol <- as.character(smoking_all_cnv$GeneSymbol)
#colnames(smoking_all_cnv)[2] <- "smoking_all_cnv"

egfr_all_cnv <- read.table("./comparison_results_1_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_EGFR_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
egfr_all_cnv$GeneSymbol <- as.character(egfr_all_cnv$GeneSymbol)
colnames(egfr_all_cnv)[2] <- "egfr_all_cnv"

#kras_all_cnv <- read.table("./comparison_results_1_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_KRAS_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
#kras_all_cnv$GeneSymbol <- as.character(kras_all_cnv$GeneSymbol)
#colnames(kras_all_cnv)[2] <- "kras_all_cnv"

#tp53_all_cnv <- read.table("./comparison_results_1_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_TP53_up_1__0____sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_all_cnv$GeneSymbol <- as.character(tp53_all_cnv$GeneSymbol)
#colnames(tp53_all_cnv)[2] <- "tp53_all_cnv"

#egfr_female_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_EGFR_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#egfr_female_cnv$GeneSymbol <- as.character(egfr_female_cnv$GeneSymbol)
#colnames(egfr_female_cnv)[2] <- "egfr_female_cnv"

egfr_male_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_EGFR_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_male_cnv$GeneSymbol <- as.character(egfr_male_cnv$GeneSymbol)
colnames(egfr_male_cnv)[2] <- "egfr_male_cnv"

#kras_female_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_KRAS_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#kras_female_cnv$GeneSymbol <- as.character(kras_female_cnv$GeneSymbol)
#colnames(kras_female_cnv)[2] <- "kras_female_cnv"

#kras_male_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_KRAS_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
#kras_male_cnv$GeneSymbol <- as.character(kras_male_cnv$GeneSymbol)
#colnames(kras_male_cnv)[2] <- "kras_male_cnv"

#tp53_female_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_TP53_up_female_1_female_0___sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_female_cnv$GeneSymbol <- as.character(tp53_female_cnv$GeneSymbol)
#colnames(tp53_female_cnv)[2] <- "tp53_female_cnv"

tp53_male_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Gender_TP53_up_male_1_male_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_male_cnv$GeneSymbol <- as.character(tp53_male_cnv$GeneSymbol)
colnames(tp53_male_cnv)[2] <- "tp53_male_cnv"

#egfr_asian_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_EGFR_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
#egfr_asian_cnv$GeneSymbol <- as.character(egfr_asian_cnv$GeneSymbol)
#colnames(egfr_asian_cnv)[2] <- "egfr_asian_cnv"

egfr_western_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_EGFR_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
egfr_western_cnv$GeneSymbol <- as.character(egfr_western_cnv$GeneSymbol)
colnames(egfr_western_cnv)[2] <- "egfr_western_cnv"

#gender_asian_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_Gender_up_asian_female_asian_male___sig_outliers.txt", header = TRUE, sep = "\t")
#gender_asian_cnv$GeneSymbol <- as.character(gender_asian_cnv$GeneSymbol)
#colnames(gender_asian_cnv)[2] <- "gender_asian_cnv"

#region_female_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_Gender_up_asian_female_western_female___sig_outliers.txt", header = TRUE, sep = "\t")
#region_female_cnv$GeneSymbol <- as.character(region_female_cnv$GeneSymbol)
#colnames(region_female_cnv)[2] <- "region_female_cnv"

region_male_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_Gender_up_asian_male_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
region_male_cnv$GeneSymbol <- as.character(region_male_cnv$GeneSymbol)
colnames(region_male_cnv)[2] <- "region_male_cnv"

#gender_western_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_Gender_up_western_female_western_male___sig_outliers.txt", header = TRUE, sep = "\t")
#gender_western_cnv$GeneSymbol <- as.character(gender_western_cnv$GeneSymbol)
#colnames(gender_western_cnv)[2] <- "gender_western_cnv"

#kras_asian_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_KRAS_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
#kras_asian_cnv$GeneSymbol <- as.character(kras_asian_cnv$GeneSymbol)
#colnames(kras_asian_cnv)[2] <- "kras_asian_cnv"

#kras_western_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_KRAS_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
#kras_western_cnv$GeneSymbol <- as.character(kras_western_cnv$GeneSymbol)
#colnames(kras_western_cnv)[2] <- "kras_western_cnv"

#tp53_asian_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_TP53_up_asian_1_asian_0___sig_outliers.txt", header = TRUE, sep = "\t")
#tp53_asian_cnv$GeneSymbol <- as.character(tp53_asian_cnv$GeneSymbol)
#colnames(tp53_asian_cnv)[2] <- "tp53_asian_cnv"

tp53_western_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Region_TP53_up_western_1_western_0___sig_outliers.txt", header = TRUE, sep = "\t")
tp53_western_cnv$GeneSymbol <- as.character(tp53_western_cnv$GeneSymbol)
colnames(tp53_western_cnv)[2] <- "tp53_western_cnv"

#smoker_female_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_Gender_up_smoker_female_non-smoker_female___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_female_cnv$GeneSymbol <- as.character(smoker_female_cnv$GeneSymbol)
#colnames(smoker_female_cnv)[2] <- "smoker_female_cnv"

#smoker_male_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_Gender_up_smoker_male_non-smoker_male___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_male_cnv$GeneSymbol <- as.character(smoker_male_cnv$GeneSymbol)
#colnames(smoker_male_cnv)[2] <- "smoker_male_cnv"

#smoker_asian_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_Region_up_smoker_asian_non-smoker_asian___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_asian_cnv$GeneSymbol <- as.character(smoker_asian_cnv$GeneSymbol)
#colnames(smoker_asian_cnv)[2] <- "smoker_asian_cnv"

#smoker_western_cnv<- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_Region_up_smoker_western_non-smoker_western___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_western_cnv$GeneSymbol <- as.character(smoker_western_cnv$GeneSymbol)
#colnames(smoker_western_cnv)[2] <- "smoker_western_cnv"

#smoker_egfr_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_EGFR_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_egfr_cnv$GeneSymbol <- as.character(smoker_egfr_cnv$GeneSymbol)
#colnames(smoker_egfr_cnv)[2] <- "smoker_egfr_cnv"

#smoker_kras_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_KRAS_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_kras_cnv$GeneSymbol <- as.character(smoker_kras_cnv$GeneSymbol)
#colnames(smoker_kras_cnv)[2] <- "smoker_kras_cnv"

#smoker_tp53_cnv <- read.table("./comparison_results_2_dgidb_fdr_0.01_iqr_1.25/luad2_cnv_dgidb_0.01_iqr_1.25_Smoking_TP53_up_smoker_1_non-smoker_1___sig_outliers.txt", header = TRUE, sep = "\t")
#smoker_tp53_cnv$GeneSymbol <- as.character(smoker_tp53_cnv$GeneSymbol)
#colnames(smoker_tp53_cnv)[2] <- "smoker_tp53_cnv"

###

#cnv_dgidb <- full_join(egfr_all_cnv, egfr_female_cnv, by = "GeneSymbol")
cnv_dgidb <- full_join(egfr_all_cnv, egfr_male_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, egfr_asian_cnv, by = "GeneSymbol")
cnv_dgidb <- full_join(cnv_dgidb, egfr_western_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, kras_all_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, kras_female_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, kras_male_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, kras_asian_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, kras_western_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, tp53_all_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, tp53_female_cnv, by = "GeneSymbol")
cnv_dgidb <- full_join(cnv_dgidb, tp53_male_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, tp53_asian_cnv, by = "GeneSymbol")
cnv_dgidb <- full_join(cnv_dgidb, tp53_western_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, region_all_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, region_female_cnv, by = "GeneSymbol")
cnv_dgidb <- full_join(cnv_dgidb, region_male_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, gender_all_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, gender_asian_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, gender_western_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoking_all_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_female_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_male_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_asian_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_western_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_egfr_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_tp53_cnv, by = "GeneSymbol")
#cnv_dgidb <- full_join(cnv_dgidb, smoker_kras_cnv, by = "GeneSymbol")

cnv_dgidb[is.na(cnv_dgidb)] <- 0

##

cnv_dgidb_gene <- cnv_dgidb[,-1]
rownames(cnv_dgidb_gene) <- cnv_dgidb[,1]
cohorts_all_cnv <- colnames(cnv_dgidb_gene)
cnv_dgidb_annotation <- HeatmapAnnotation(df = as.data.frame(cohorts_all_cnv),
                                       col = list(cohorts_all_cnv = c(egfr_all_cnv = "coral",
                                                              #egfr_female_cnv = "dodgerblue4",
                                                              egfr_male_cnv = "dodgerblue1",
                                                              #egfr_asian_cnv= "deepskyblue4",
                                                              egfr_western_cnv= "deepskyblue2",
                                                              #kras_all_cnv = "coral",
                                                              #kras_female_cnv = "mediumorchid4",
                                                              #kras_male_cnv = "mediumorchid1",
                                                              #kras_asian_cnv= "mediumpurple4",
                                                              #kras_western_cnv= "mediumpurple1",
                                                              #tp53_all_cnv = "coral",
                                                              #tp53_female_cnv = "olivedrab4",
                                                              tp53_male_cnv = "olivedrab3",
                                                              #tp53_asian_cnv= "darkgreen",
                                                              tp53_western_cnv= "springgreen4",
                                                              #region_all_cnv = "coral",
                                                              #region_female_cnv = "darkorange3",
                                                              region_male_cnv = "darkorange")))
                                                              #gender_all_cnv = "coral",
                                                              #gender_asian_cnv= "darkgoldenrod3",
                                                              #gender_western_cnv= "darkgoldenrod1",
                                                              #smoking_all_cnv = "coral",
                                                              #smoker_female_cnv = "lightsteelblue4",
                                                              #smoker_male_cnv = "lightsteelblue1",
                                                              #smoker_asian_cnv= "mistyrose4",
                                                              #smoker_western_cnv= "mistyrose2",
                                                              #smoker_egfr_cnv = "aquamarine4",
                                                              #smoker_kras_cnv = "aquamarine3",
                                                              #smoker_tp53_cnv = "aquamarine1")))

cnv_dgidb_heat <- Heatmap(cnv_dgidb_gene,
                         top_annotation = cnv_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 5),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(3, "in"),
                                                     title = NULL),
                         col = colorRamp2(c(0, 1.0e-50, 0.005, 0.05), 
                                          c("gray85", "red3", "orangered2", "oldlace")),
                         row_title = "Q Value",
                         row_title_gp = gpar(fontsize = 10))
  
cnv_dgidb_heat_2 <- draw(cnv_dgidb_heat, heatmap_legend_side = "left")

```


```{r cnv dgidb query}

# DGIDB
cnv_dgidb_gene

# Filterd
druggable_dgidb_cnv <- rownames(cnv_dgidb_gene)
length(druggable_dgidb_cnv) #63
druggable_dgidb_cnv

# Query 
result_drug_dgidb_cnv <- queryDGIdb(druggable_dgidb_cnv)

result_drug_dgidb_cnv

```


```{r cnv dgidb result summary}

## Result summary
# Score is the total number of source databases listing the interaction (including PubMed articles)
summary_drug_dgidb_cnv <- resultSummary(result_drug_dgidb_cnv) 
summary_drug_dgidb_score_cnv <- dplyr::select(summary_drug_dgidb_cnv, Gene, Score)

# Change to numeric and aggregate for total score
summary_drug_dgidb_score_cnv$Score <- as.numeric(summary_drug_dgidb_score_cnv$Score)
summary_drug_dgidb_score_cnv <- aggregate(. ~ Gene, data = summary_drug_dgidb_score_cnv, sum)

# Get log score
summary_drug_dgidb_score_cnv$Log_Score <- log(summary_drug_dgidb_score_cnv$Score)

# Order rows to match heatmap
target_cnv <- as.vector(rownames(cnv_dgidb_gene))
target_cnv
summary_drug_dgidb_score_cnv <- summary_drug_dgidb_score_cnv[match(target_cnv, summary_drug_dgidb_score_cnv$Gene),]

# Label row names for heatmap
row.names(summary_drug_dgidb_score_cnv) <- target_cnv
summary_drug_dgidb_score_cnv

# Replace NA with 0 for plotting
summary_drug_dgidb_score_cnv$Score[is.na(summary_drug_dgidb_score_cnv$Score) == TRUE] <- 0
summary_drug_dgidb_score_cnv$Log_Score[is.na(summary_drug_dgidb_score_cnv$Log_Score) == TRUE] <- 0


summary_drug_dgidb_score_cnv

```


```{r cnv import interactions and aggregate}

# Import manual interactions then aggregate and separate into 1/0
interactions_cnv <- read.delim("./druggable/DGIDB_gene_drug_interactions_20190417_modified_dup_removed.txt", sep = "\t")
interactions_cnv <- interactions_cnv[interactions_cnv$gene_name %in% druggable_dgidb_cnv ,]

# Select gene name and interaction type only
interactions_cnv <- dplyr::select(interactions_cnv, gene_name, interaction_types)

# Aggregate to a list
interactions_cnv <- aggregate(interaction_types ~ gene_name, interactions_cnv, paste, collapse = ",")

# Order rows to match heatmap
target_int_cnv <- as.vector(druggable_dgidb_cnv)
interactions_cnv <- interactions_cnv[match(target_int_cnv, interactions_cnv$gene_name),]

interactions_cnv

```


```{r cnv manual interactions data frame}

# Rename columns
colnames(interactions_cnv)[1:2] <- c("gene", "interaction_type")


# if interaction_type includes this exact match, put 1 else put 0 

interactions_cnv <- separate(interactions_cnv, interaction_type, 
                         into=c("int_1", "int_2", "int_3", "int_4", "int_5", "int_6", "int_7"), sep="," )
interactions_cnv$int_1[interactions_cnv$int_1==''] <- NA
interactions_cnv$int_1[is.na(interactions_cnv$int_1) == TRUE] <- 0
interactions_cnv$int_2[is.na(interactions_cnv$int_2) == TRUE] <- 0
interactions_cnv$int_3[is.na(interactions_cnv$int_3) == TRUE] <- 0
interactions_cnv$int_4[is.na(interactions_cnv$int_4) == TRUE] <- 0
interactions_cnv$int_5[is.na(interactions_cnv$int_5) == TRUE] <- 0
interactions_cnv$int_6[is.na(interactions_cnv$int_6) == TRUE] <- 0
interactions_cnv$int_7[is.na(interactions_cnv$int_7) == TRUE] <- 0

interactions_cnv

interactions_cnv$Antagonist <-ifelse((interactions_cnv$int_1 == "antagonist" |
                                interactions_cnv$int_2 == "antagonist" |
                                interactions_cnv$int_3 == "antagonist" |
                                interactions_cnv$int_4 == "antagonist" |
                                interactions_cnv$int_5 == "antagonist" |
                                interactions_cnv$int_6 == "antagonist" |
                                interactions_cnv$int_7== "antagonist"), 1, 0)

interactions_cnv$Inhibitor <-ifelse((interactions_cnv$int_1 == "inhibitor" |
                                interactions_cnv$int_1 == "gating inhibitor" |
                                interactions_cnv$int_2 == "inhibitor" |
                                interactions_cnv$int_2 == "gating inhibitor" |
                                interactions_cnv$int_3 == "inhibitor" |
                                interactions_cnv$int_3 == "gating inhibitor" |
                                interactions_cnv$int_4 == "inhibitor" |
                                interactions_cnv$int_4 == "gating inhibitor" |
                                interactions_cnv$int_5 == "inhibitor"  |
                                interactions_cnv$int_5 == "gating inhibitor"|
                                interactions_cnv$int_6 == "inhibitor" |
                                interactions_cnv$int_6 == "gating inhibitor" |
                                interactions_cnv$int_7 == "inhibitor" |
                                interactions_cnv$int_7 == "gating inhibitor"), 1, 0)

interactions_cnv$Agonist <-ifelse((interactions_cnv$int_1 == "agonist"|
                                interactions_cnv$int_1 == "partial agonist"|
                                interactions_cnv$int_1 == "inverse agonist" |
                                interactions_cnv$int_2 == "agonist"|
                                interactions_cnv$int_2 == "partial agonist"|
                                interactions_cnv$int_2 == "inverse agonist" |
                                interactions_cnv$int_3 == "agonist"|
                                interactions_cnv$int_3 == "partial agonist"|
                                interactions_cnv$int_3 == "inverse agonist" |
                                interactions_cnv$int_4 == "agonist"|
                                interactions_cnv$int_4 == "partial agonist"|
                                interactions_cnv$int_4 == "inverse agonist" |
                                interactions_cnv$int_5 == "agonist"|
                                interactions_cnv$int_5 == "partial agonist"|
                                interactions_cnv$int_5 == "inverse agonist" |
                                interactions_cnv$int_6 == "agonist" |
                                interactions_cnv$int_6 == "partial agonist"|
                                interactions_cnv$int_6 == "inverse agonist" |
                                interactions_cnv$int_7 == "agonist" |
                                interactions_cnv$int_7 == "partial agonist"|
                                interactions_cnv$int_7 == "inverse agonist"), 1, 0)

interactions_cnv$Antibody <-ifelse((interactions_cnv$int_1 == "antibody" |
                                interactions_cnv$int_2 == "antibody" |
                                interactions_cnv$int_3 == "antibody" |
                                interactions_cnv$int_4 == "antibody" |
                                interactions_cnv$int_5 == "antibody" |
                                interactions_cnv$int_6 == "antibody" |
                                interactions_cnv$int_7 == "antibody"), 1, 0)

interactions_cnv$Activator <-ifelse((interactions_cnv$int_1 == "activator" |
                                interactions_cnv$int_2 == "activator" |
                                interactions_cnv$int_3 == "activator" |
                                interactions_cnv$int_4 == "activator" |
                                interactions_cnv$int_5 == "activator" |
                                interactions_cnv$int_6 == "activator" |
                                interactions_cnv$int_7 == "activator"), 1, 0)

interactions_cnv$Binder <-ifelse((interactions_cnv$int_1 == "binder" |
                                interactions_cnv$int_2 == "binder" |
                                interactions_cnv$int_3 == "binder" |
                                interactions_cnv$int_4 == "binder" |
                                interactions_cnv$int_5 == "binder" |
                                interactions_cnv$int_6 == "binder" |
                                interactions_cnv$int_7 == "binder"), 1, 0)

interactions_cnv$Blocker <-ifelse((interactions_cnv$int_1 == "blocker" | 
                                 interactions_cnv$int_1 == "channel blocker" |
                                interactions_cnv$int_2 == "blocker" | 
                                 interactions_cnv$int_2 == "channel blocker" |
                                interactions_cnv$int_3 == "blocker" | 
                                 interactions_cnv$int_3 == "channel blocker" |
                                interactions_cnv$int_4 == "blocker" | 
                                 interactions_cnv$int_4 == "channel blocker" |
                                interactions_cnv$int_5 == "blocker" | 
                                 interactions_cnv$int_5 == "channel blocker" |
                                interactions_cnv$int_6 == "blocker" | 
                                 interactions_cnv$int_6 == "channel blocker" |
                                interactions_cnv$int_7 == "blocker" | 
                                 interactions_cnv$int_7 == "channel blocker"), 1, 0)

#allosteric modulator

other_list <- c("antisense", "vaccine", "stimulator", "positive allosteric modulator",
                "allosteric modulator", "cofactor", "antisense oligonucleotide")

interactions_cnv$Other <- ifelse(((interactions_cnv$int_1 %in% other_list) == TRUE) |
  ((interactions_cnv$int_2 %in% other_list) == TRUE) |
  ((interactions_cnv$int_3 %in% other_list) == TRUE) |
  ((interactions_cnv$int_4 %in% other_list) == TRUE) |
  ((interactions_cnv$int_5 %in% other_list) == TRUE) |
  ((interactions_cnv$int_6 %in% other_list) == TRUE) |
  ((interactions_cnv$int_7 %in% other_list) == TRUE), 1, 0)


interactions_cnv_final <- dplyr::select(interactions_cnv, -int_1, -int_2, -int_3, -int_4, 
                                        -int_5, -int_6, -int_7)
interactions_cnv_final


# Aggregate and sum by gene name 
interactions_cnv_agg <- aggregate(. ~ gene, data = interactions_cnv_final, max)

# Order rows to match heatmap
target_int_cnv <- as.vector(druggable_dgidb_cnv)
interactions_cnv_agg <- interactions_cnv_agg[match(target_int_cnv, interactions_cnv_agg$gene),]

# Label row names for heatmap
row.names(interactions_cnv_agg) <- target_int_cnv
interactions_cnv_agg

# Remove gene column
interactions_cnv_agg <- dplyr::select(interactions_cnv_agg, -gene)

# Replace NA with 0 for plotting
interactions_cnv_agg[is.na(interactions_cnv_agg)] <- 0

interactions_cnv_agg$Inhibitor[interactions_cnv_agg$Inhibitor == 1] <- "Inhibitor"
interactions_cnv_agg$Antagonist[interactions_cnv_agg$Antagonist == 1] <- "Antagonist"
interactions_cnv_agg$Agonist[interactions_cnv_agg$Agonist == 1] <- "Agonist"
interactions_cnv_agg$Antibody[interactions_cnv_agg$Antibody == 1] <- "Antibody"
interactions_cnv_agg$Activator[interactions_cnv_agg$Activator == 1] <- "Activator"
interactions_cnv_agg$Binder[interactions_cnv_agg$Binder == 1] <- "Binder"
interactions_cnv_agg$Blocker[interactions_cnv_agg$Blocker == 1] <- "Blocker"
interactions_cnv_agg$Other[interactions_cnv_agg$Other == 1] <- "Other"

interactions_cnv_agg[interactions_cnv_agg == 0 ] <- "NA"

interactions_cnv_agg

```


```{r cnv fda approved drugs}

# FDA approved drugs 
fda_cnv <- read.table("./druggable/druggable_FDA_approved_2019-04-17_list.txt")
fda_cnv$FDA_approved <- "True"
colnames(fda_cnv)[1] <- c("gene")

# Intersect with druggable genes
fda_int_cnv <- as.data.frame(row.names(interactions_cnv_agg))
colnames(fda_int_cnv) <- c("gene")
fda_int_cnv <- left_join(fda_int_cnv, fda_cnv, by = "gene")

# Change NA to 
fda_int_cnv$FDA_approved[is.na(fda_int_cnv$FDA_approved) == TRUE] <- "False"
fda_int_cnv

```


```{r cnv heatmap}

### Final Figure Heatmap: Druggable Filtered ###`

# Druggable Heatmap
cnv_drug_dgidb_annotation <- HeatmapAnnotation(Comparison = c(rep('EGFR', 3),
                                                    rep('TP53', 2), rep('Region', 1)), 
                                       Cohort = c("All", "Male", "Western",
                                                  "Male", "Western", 
                                                  "Male"),
                                       col = list(Comparison = c(EGFR = "#EF7C12",
                                                              #KRAS = "#FFB90F",
                                                              TP53 = "#F7E690",
                                                              Region = "#EFC7E6"),
                                                              #Gender = "#CB87B4",
                                                              #Smoker = "#B25D91"),
                                                  Cohort = c(All = "Black",
                                                             #Female = "#AFDFEF",
                                                             Male = "#54BCD1" ,
                                                              #Asian = "#0076C0",
                                                              Western = "#27408B")),
                                                            #EGFR_Mutation = "#EDA9AB",
                                                              #TP53_Mutation = "#D9565C")),
                                       annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                                   legend_width = unit(0.20, "cm"),
                                                                   title_gp = gpar(fontsize = 11),
                                                                   labels_gp = gpar(fontsize = 9)),
                                       name = "cnv_heat_anno")

cnv_heat_drug_dgidb <- Heatmap(cnv_dgidb_gene,
                                name = "cnv_heat",
                         top_annotation = cnv_drug_dgidb_annotation,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 5),
                         top_annotation_height = unit(1, "cm"),
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Q Value",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colorRamp2(c(0, 1.0e-50, 0.01, 0.05), 
                                          c("gray90", "red3", "orangered2", "oldlace")),
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75))

# Decorate heatmap
cnv_heat_drug_dgidb_fig <- draw(cnv_heat_drug_dgidb, heatmap_legend_side = "right")


# Creat heatmap color scheme
colors <-  structure(c("#3C5541", "#3C8C4D", "#3CBC38", 
                       "#AFDE62", "#DFCEE0", "#CDB5CD",
                       "#7473A6", "#132157", "gray90"), 
                     names = c("Antagonist", "Inhibitor", "Agonist", "Antibody", 
                               "Activator", "Binder",
                               "Blocker", "Other", "NA"))

# Heatmap annotations 
ha_rot_cn_dgidb <- HeatmapAnnotation(Interaction = c("Antagonist","Inhibitor","Agonist", "Antibody", 
                                                     "Activator", "Binder", "Blocker", "Other"),
                               col = list(Interaction = c(Antagonist = "#3C5541",
                                                          Inhibitor = "#3C8C4D",
                                                          Agonist = "#3CBC38",
                                                          Antibody = "#AFDE62",
                                                          Activator = "#DFCEE0",
                                                          Binder = "#CDB5CD",
                                                          Blocker = "#7473A6",
                                                          Other = "#132157")),
                               annotation_legend_param = list(legend_height = unit(1, "cm"),
                                                              title = "Interaction",
                                                              legend_width = unit(0.20, "cm"),
                                                              title_gp = gpar(fontsize = 11),
                                                              labels_gp = gpar(fontsize = 9)),
                               show_legend = FALSE)


# Barplot
row_anno_dgidb_cnv <- rowAnnotation(barplot = row_anno_barplot(summary_drug_dgidb_score_cnv$Log_Score,
                          axis_direction = "reverse", border = TRUE, axis = TRUE, axis_side = "top",
                          gp = gpar(fill = "black")), 
                          width = unit(2, "cm"))

                             
# Create heatmap
interactions_cnv_drug_dgidb_heat <- Heatmap(interactions_cnv_agg,
                         show_column_names = F, 
                         show_row_names = T,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "Interaction",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors,
                         row_title_gp = gpar(fontsize = 10),
                         column_names_side = "top",
                         top_annotation = ha_rot_cn_dgidb,
                         width = unit(2, "cm"),
                         rect_gp = gpar(col = "white", lwd = 0.75),
                         name = "cnv_int")

# Colors for FDA heatmap
colors_2 <-  structure(c("#007BC3", "#AFDFEF"), 
                     names = c("True", "False"))


# Create FDA heatmap
fda_drug_dgidb_heat_cnv <- Heatmap(fda_int_cnv$FDA_approved,
                         show_column_names = F, 
                         show_row_names = F,
                         cluster_rows = F,
                         cluster_columns = F,
                         row_names_gp = gpar(fontsize = 0),
                         column_names_gp = gpar(fontsize = 2),
                         top_annotation_height = unit(1, "cm"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(legend_height = unit(1, "cm"),
                                                     legend_width = unit(0.20, "cm"),
                                                     title = "FDA Approved",
                                                     title_gp = gpar(fontsize = 11),
                                                     labels_gp = gpar(fontsize = 9)),
                         col = colors_2,
                         row_title_gp = gpar(fontsize = 10),
                         rect_gp = gpar(col = "white", lwd = 0.75),
                         name = "cnv_fda")


#detail_drug_heat
lgd <- Legend(labels = "Score", title = "Druggability", 
              legend_gp = gpar(fill = "black"),
              title_gp = gpar(fontsize = 11),
              labels_gp = gpar(fontsize = 9))

```


```{r cnv heatmap decorations}

final_druggable_cnv <- draw(row_anno_dgidb_cnv + fda_drug_dgidb_heat_cnv +
                          interactions_cnv_drug_dgidb_heat + cnv_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("cnv_heat", {
	grid.rect(0, 0, width = 3/6, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_heat", {
	grid.rect(3/6, 0, width = 2/6, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_heat", {
	grid.rect(5/6, 0, width = 1/6, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_int", {
	grid.rect(0, 0, width = 1, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_int", {
	grid.rect(0, 0, width = 1, height = 1.11, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_fda", {
	grid.rect(0, 0, width = 1, height = 1, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})


```


```{r cnv heatmap save}

png("cnv_dgidb_final_druggable_heatmap.png", units="in", width=6, height=6, res=800)

final_druggable_cnv <- draw(row_anno_dgidb_cnv + fda_drug_dgidb_heat_cnv +
                          interactions_cnv_drug_dgidb_heat + cnv_heat_drug_dgidb_fig,
                        gap = unit(c(2, 2, 0, 0), "mm"), annotation_legend_list = list(lgd))

decorate_heatmap_body("cnv_heat", {
	grid.rect(0, 0, width = 3/6, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_heat", {
	grid.rect(3/6, 0, width = 2/6, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_heat", {
	grid.rect(5/6, 0, width = 1/6, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_int", {
	grid.rect(0, 0, width = 1, height = 1.08, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

decorate_heatmap_body("cnv_fda", {
	grid.rect(0, 0, width = 1, height = 1, just = c("left", "bottom"), 
		gp = gpar(lwd = 0.75, col = "black", fill = "transparent"))
})

dev.off()

```


---





---



---